<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>NetCoMi  • NetCoMi</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png"><script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="NetCoMi "><meta property="og:image" content="https://netcomi.de/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">NetCoMi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="articles/NetCoMi.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="articles/net_comparison.html">Network comparison</a></li>
    <li><a class="dropdown-item" href="articles/diffnet.html">Differential networks</a></li>
    <li><a class="dropdown-item" href="articles/dissimilarity_networks.html">Dissimilarity-based Networks</a></li>
    <li><a class="dropdown-item" href="articles/create_asso_perm.html">Generate permuted association matrices</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>More examples</h6></li>
    <li><a class="dropdown-item" href="articles/asso_as_input.html">Association matrix as input</a></li>
    <li><a class="dropdown-item" href="articles/cross_domain_networks.html">Cross-domain networks</a></li>
    <li><a class="dropdown-item" href="articles/soil_example.html">Soil microbiome example</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/stefpeschel/NetCoMi/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="logo.png" class="logo" alt=""><h1>NetCoMi </h1>
      <small class="dont-index">Source: <a href="https://github.com/stefpeschel/NetCoMi/blob/main/readme.md" class="external-link"><code>readme.md</code></a></small>
    </div>

<div id="netcomi-" class="section level1">

<p><a href="https://zenodo.org/badge/latestdoi/259906607" class="external-link"><img src="https://zenodo.org/badge/259906607.svg" alt="DOI"></a> <a href="https://anaconda.org/bioconda/r-netcomi" class="external-link"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda"></a></p>
<p>NetCoMi (<strong>Net</strong>work <strong>Co</strong>nstruction and Comparison for <strong>Mi</strong>crobiome Data) provides functionality for constructing, analyzing, and comparing networks suitable for the application on microbial compositional data. The R package implements the workflow proposed in</p>
<p>Stefanie Peschel, Christian L Müller, Erika von Mutius, Anne-Laure Boulesteix, Martin Depner (2020). <a href="https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbaa290/6017455" class="external-link">NetCoMi: network construction and comparison for microbiome data in R</a>. <em>Briefings in Bioinformatics</em>, bbaa290. <a href="https://doi.org/10.1093/bib/bbaa290" class="external-link uri">https://doi.org/10.1093/bib/bbaa290</a>.</p>
<p>NetCoMi allows its users to construct, analyze, and compare microbial association or dissimilarity networks in a fast and reproducible manner. Starting with a read count matrix originating from a sequencing process, the pipeline includes a wide range of existing methods for treating zeros in the data, normalization, computing microbial associations or dissimilarities, and sparsifying the resulting association/ dissimilarity matrix. These methods can be combined in a modular fashion to generate microbial networks. NetCoMi can either be used for constructing, analyzing and visualizing a single network, or for comparing two networks in a graphical as well as a quantitative manner, including statistical tests. The package furthermore offers functionality for constructing differential networks, where only differentially associated taxa are connected.</p>
<p><img src="reference/figures/networkplot_soil.png" width="800"></p>
<blockquote>
<p>Exemplary network comparison using soil microbiome data (<a href="https://github.com/joey711/phyloseq/blob/master/data/soilrep.RData" class="external-link">‘soilrep’ data from phyloseq package</a>). Microbial associations are compared between the two experimantal settings ‘warming’ and ‘non-warming’ using the same layout in both groups.</p>
</blockquote>
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a></h2>
<ol style="list-style-type: decimal"><li><a href="#methods-included-in-NetCoMi">Methods included in NetCoMi</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#development-version">Development version</a></li>
<li>
<a href="#usage">Usage</a>
<ul><li><a href="#network-with-spring-as-association-measure">Network with SPRING as association measure</a></li>
<li><a href="#export-to-gephi">Export to Gephi</a></li>
<li><a href="#network-with-pearson-correlation-as-association-measure">Network with Pearson correlations</a></li>
<li><a href="#using-the-unsigned-transformation">“Unsigned” transformation</a></li>
<li><a href="#network-on-genus-level">Network on genus level</a></li>
<li><a href="#using-an-association-matrix-as-input">Association matrix as input</a></li>
<li><a href="#network-comparison">Network comparison</a></li>
<li><a href="#differential-networks">Differential networks</a></li>
<li><a href="#dissimilarity-based-networks">Dissimilarity-based Networks</a></li>
<li><a href="#soil-microbiome-example">Soil microbiome example</a></li>
</ul></li>
<li><a href="#references">References</a></li>
</ol></div>
<div class="section level2">
<h2 id="methods-included-in-netcomi">Methods included in NetCoMi<a class="anchor" aria-label="anchor" href="#methods-included-in-netcomi"></a></h2>
<p>Here is an overview of methods available for network construction, together with some information on their implementation in R:</p>
<p><strong>Association measures:</strong></p>
<ul><li>Pearson coefficient (<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/cor" class="external-link"><code>cor()</code></a> from <code>stats</code> package)</li>
<li>Spearman coefficient (<a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/cor" class="external-link"><code>cor()</code></a> from <code>stats</code> package)</li>
<li>Biweight Midcorrelation <a href="https://rdrr.io/cran/WGCNA/man/bicor.html" class="external-link"><code>bicor()</code></a> from <code>WGCNA</code> package</li>
<li>SparCC (<a href="https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html" class="external-link"><code>sparcc()</code></a> from <code>SpiecEasi</code> package)</li>
<li>CCLasso (<a href="https://github.com/huayingfang/CCLasso" class="external-link">R code on GitHub</a>)</li>
<li>CCREPE (<a href="https://bioconductor.org/packages/release/bioc/html/ccrepe.html" class="external-link"><code>ccrepe</code></a> package)</li>
<li>SpiecEasi (<a href="https://github.com/zdk123/SpiecEasi" class="external-link"><code>SpiecEasi</code></a> package)</li>
<li>SPRING (<a href="https://github.com/GraceYoon/SPRING" class="external-link"><code>SPRING</code></a> package)</li>
<li>gCoda (<a href="https://github.com/huayingfang/gCoda" class="external-link">R code on GitHub</a>)</li>
<li>propr (<a href="https://cran.r-project.org/web/packages/propr/index.html" class="external-link"><code>propr</code></a> package)</li>
</ul><p><strong>Dissimilarity measures:</strong></p>
<ul><li>Euclidean distance (<a href="https://www.rdocumentation.org/packages/vegan/versions/2.4-2/topics/vegdist" class="external-link"><code>vegdist()</code></a> from <code>vegan</code> package)</li>
<li>Bray-Curtis dissimilarity (<a href="https://www.rdocumentation.org/packages/vegan/versions/2.4-2/topics/vegdist" class="external-link"><code>vegdist()</code></a> from <code>vegan</code> package)</li>
<li>Kullback-Leibler divergence (KLD) (<a href="https://rdrr.io/cran/LaplacesDemon/man/KLD.html" class="external-link"><code>KLD()</code></a> from <code>LaplacesDemon</code> package)</li>
<li>Jeffrey divergence (own code using <a href="https://rdrr.io/cran/LaplacesDemon/man/KLD.html" class="external-link"><code>KLD()</code></a> from <code>LaplacesDemon</code> package)</li>
<li>Jensen-Shannon divergence (own code using <a href="https://rdrr.io/cran/LaplacesDemon/man/KLD.html" class="external-link"><code>KLD()</code></a> from <code>LaplacesDemon</code> package)</li>
<li>Compositional KLD (own implementation following Martin-Fernández et al. (1999))</li>
<li>Aitchison distance (<a href="https://www.rdocumentation.org/packages/vegan/versions/2.4-2/topics/vegdist" class="external-link"><code>vegdist()</code></a> and <a href="https://rdrr.io/github/zdk123/SpiecEasi/man/clr.html" class="external-link"><code>clr()</code></a> from <code>SpiecEasi</code> package)</li>
</ul><p><strong>Methods for zero replacement:</strong></p>
<ul><li>Add a predefined pseudo count to the count table</li>
<li>Replace only zeros in the count table by a predefined pseudo count (ratios between non-zero values are preserved)</li>
<li>Multiplicative replacement (<a href="https://rdrr.io/cran/zCompositions/man/multRepl.html" class="external-link"><code>multRepl</code></a> from <code>zCompositions</code> package)</li>
<li>Modified EM alr-algorithm (<a href="https://rdrr.io/cran/zCompositions/man/lrEM.html" class="external-link"><code>lrEM</code></a> from <code>zCompositions</code> package)</li>
<li>Bayesian-multiplicative replacement (<a href="https://rdrr.io/cran/zCompositions/man/cmultRepl.html" class="external-link"><code>cmultRepl</code></a> from <code>zCompositions</code> package)</li>
</ul><p><strong>Normalization methods:</strong></p>
<ul><li>Total Sum Scaling (TSS) (own implementation)</li>
<li>Cumulative Sum Scaling (CSS) (<a href=""><code>cumNormMat</code></a> from <code>metagenomeSeq</code> package)</li>
<li>Common Sum Scaling (COM) (own implementation)</li>
<li>Rarefying (<a href=""><code>rrarefy</code></a> from <code>vegan</code> package)</li>
<li>Variance Stabilizing Transformation (VST) (<a href=""><code>varianceStabilizingTransformation</code></a> from <code>DESeq2</code> package)</li>
<li>Centered log-ratio (clr) transformation (<a href="https://rdrr.io/github/zdk123/SpiecEasi/man/clr.html" class="external-link"><code>clr()</code></a> from <code>SpiecEasi</code> package))</li>
</ul><p>TSS, CSS, COM, VST, and the clr transformation are described in (Badri et al. 2020).</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a></h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Required packages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install NetCoMi</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"stefpeschel/NetCoMi"</span>, </span>
<span>                         dependencies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"LinkingTo"</span><span class="op">)</span>,</span>
<span>                         repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://cloud.r-project.org/"</span>,</span>
<span>                                   <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>If there are any errors during installation, please install the missing dependencies manually.</p>
<p>In particular the automatic installation of <a href="https://github.com/GraceYoon/SPRING" class="external-link"><code>SPRING</code></a> and <a href="https://github.com/zdk123/SpiecEasi" class="external-link"><code>SpiecEasi</code></a> (only available on GitHub) does sometimes not work. These packages can be installed as follows (the order is important because SPRING depends on SpiecEasi):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"zdk123/SpiecEasi"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"GraceYoon/SPRING"</span><span class="op">)</span></span></code></pre></div>
<p>Packages that are optionally required in certain settings are not installed together with NetCoMi. These can be installed automatically using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/installNetCoMiPacks.html">installNetCoMiPacks</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Please check:</span></span>
<span><span class="op">?</span><span class="fu"><a href="reference/installNetCoMiPacks.html">installNetCoMiPacks</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>If not installed via <code><a href="reference/installNetCoMiPacks.html">installNetCoMiPacks()</a></code>, the required package is installed by the respective NetCoMi function when needed.</p>
<div class="section level3">
<h3 id="bioconda">Bioconda<a class="anchor" aria-label="anchor" href="#bioconda"></a></h3>
<p>Thanks to <a href="https://github.com/daydream-boost" class="external-link">daydream-boost</a>, NetCoMi can also be installed from conda bioconda channel with</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># You can install an individual environment firstly with</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># conda create -n NetCoMi</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># conda activate NetCoMi</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda <span class="at">-c</span> conda-forge r-netcomi</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="development-version">Development version<a class="anchor" aria-label="anchor" href="#development-version"></a></h2>
<p>Everyone who wants to use new features not included in any releases is invited to install NetCoMi’s development version:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"stefpeschel/NetCoMi"</span>, </span>
<span>                         ref <span class="op">=</span> <span class="st">"develop"</span>,</span>
<span>                         dependencies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"LinkingTo"</span><span class="op">)</span>,</span>
<span>                         repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://cloud.r-project.org/"</span>,</span>
<span>                                   <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Please check the <a href="https://github.com/stefpeschel/NetCoMi/blob/develop/NEWS.md" class="external-link">NEWS</a> document for features implemented on develop branch.</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a></h2>
<p>We use the American Gut data from <a href="https://github.com/zdk123/SpiecEasi" class="external-link"><code>SpiecEasi</code></a> package to look at some examples of how NetCoMi is applied. NetCoMi’s main functions are <code><a href="reference/netConstruct.html">netConstruct()</a></code> for network construction, <code><a href="reference/netAnalyze.html">netAnalyze()</a></code> for network analysis, and <code><a href="reference/netCompare.html">netCompare()</a></code> for network comparison. As you will see in the following, these three functions must be executed in the aforementioned order. A further function is <code><a href="reference/diffnet.html">diffnet()</a></code> for constructing a differential association network. <code><a href="reference/diffnet.html">diffnet()</a></code> must be applied to the object returned by <code><a href="reference/netConstruct.html">netConstruct()</a></code>.</p>
<p>First of all, we load NetCoMi and the data from American Gut Project (provided by <a href="https://github.com/zdk123/SpiecEasi" class="external-link"><code>SpiecEasi</code></a>, which is automatically loaded together with NetCoMi).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://netcomi.de">NetCoMi</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"amgut1.filt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"amgut2.filt.phy"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="network-with-spring-as-association-measure">Network with SPRING as association measure<a class="anchor" aria-label="anchor" href="#network-with-spring-as-association-measure"></a></h3>
<div class="section level4">
<h4 id="network-construction-and-analysis">Network construction and analysis<a class="anchor" aria-label="anchor" href="#network-construction-and-analysis"></a></h4>
<p>We firstly construct a single association network using <a href="https://github.com/GraceYoon/SPRING" class="external-link">SPRING</a> for estimating associations (conditional dependence) between OTUs.</p>
<p>The data are filtered within <code><a href="reference/netConstruct.html">netConstruct()</a></code> as follows:</p>
<ul><li>Only samples with a total number of reads of at least 1000 are included (argument <code>filtSamp</code>).</li>
<li>Only the 50 taxa with highest frequency are included (argument <code>filtTax</code>).</li>
</ul><p><code>measure</code> defines the association or dissimilarity measure, which is <code>"spring"</code> in our case. Additional arguments are passed to <code><a href="https://rdrr.io/pkg/SPRING/man/SPRING.html" class="external-link">SPRING()</a></code> via <code>measurePar</code>. <code>nlambda</code> and <code>rep.num</code> are set to 10 for a decreased execution time, but should be higher for real data. <code>Rmethod</code> is set to “approx” to estimate the correlations using a hybrid multi-linear interpolation approach proposed by Yoon, Müller, and Gaynanova (2020). This method considerably reduces the runtime while controlling the approximation error.</p>
<p>Normalization as well as zero handling is performed internally in <code><a href="https://rdrr.io/pkg/SPRING/man/SPRING.html" class="external-link">SPRING()</a></code>. Hence, we set <code>normMethod</code> and <code>zeroMethod</code> to <code>"none"</code>.</p>
<p>We furthermore set <code>sparsMethod</code> to <code>"none"</code> because <code>SPRING</code> returns a sparse network where no additional sparsification step is necessary.</p>
<p>We use the “signed” method for transforming associations into dissimilarities (argument <code>dissFunc</code>). In doing so, strongly negatively associated taxa have a high dissimilarity and, in turn, a low similarity, which corresponds to edge weights in the network plot.</p>
<p>The <code>verbose</code> argument is set to 3 so that all messages generated by <code><a href="reference/netConstruct.html">netConstruct()</a></code> as well as messages of external functions are printed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_spring</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut1.filt</span>,</span>
<span>                           filtTax <span class="op">=</span> <span class="st">"highestFreq"</span>,</span>
<span>                           filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestFreq <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>                           filtSamp <span class="op">=</span> <span class="st">"totalReads"</span>,</span>
<span>                           filtSampPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>totalReads <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>,</span>
<span>                           measure <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>                           measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nlambda<span class="op">=</span><span class="fl">10</span>, </span>
<span>                                             rep.num<span class="op">=</span><span class="fl">10</span>,</span>
<span>                                             Rmethod <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span>,</span>
<span>                           normMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           zeroMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                           sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                           seed <span class="op">=</span> <span class="fl">123456</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## Checking input arguments ... Done.</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="do">## Data filtering ...</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="do">## 77 taxa removed.</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="do">## 50 taxa and 289 samples remaining.</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">## Calculate 'spring' associations ... Registered S3 method overwritten by 'dendextend':</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">##   method     from </span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">##   rev.hclust vegan</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="do">## Registered S3 method overwritten by 'seriation':</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="do">##   method         from </span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="do">##   reorder.hclust vegan</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="do">## Done.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="analyzing-the-constructed-network">Analyzing the constructed network<a class="anchor" aria-label="anchor" href="#analyzing-the-constructed-network"></a></h4>
<p>NetCoMi’s <code><a href="reference/netAnalyze.html">netAnalyze()</a></code> function is used for analyzing the constructed network(s).</p>
<p>Here, <code>centrLCC</code> is set to <code>TRUE</code> meaning that centralities are calculated only for nodes in the largest connected component (LCC).</p>
<p>Clusters are identified using greedy modularity optimization (by <code>cluster_fast_greedy()</code> from <a href="https://igraph.org/r/" class="external-link"><code>igraph</code></a> package).</p>
<p>Hubs are nodes with an eigenvector centrality value above the empirical 95% quantile of all eigenvector centralities in the network (argument <code>hubPar</code>).</p>
<p><code>weightDeg</code> and <code>normDeg</code> are set to <code>FALSE</code> so that the degree of a node is simply defined as number of nodes that are adjacent to the node.</p>
<p>By default, a heatmap of the Graphlet Correlation Matrix (GCM) is returned (with graphlet correlations in the upper triangle and significance codes resulting from Student’s t-test in the lower triangle). See <code><a href="reference/calcGCM.html">?calcGCM</a></code> and <code><a href="reference/testGCM.html">?testGCM</a></code> for details.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_spring</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_spring</span>, </span>
<span>                           centrLCC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                           hubPar <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>                           weightDeg <span class="op">=</span> <span class="cn">FALSE</span>, normDeg <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_spring_2-1.png"><!-- --></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#?summary.microNetProps</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">props_spring</span>, numbNodes <span class="op">=</span> <span class="fl">5L</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="do">## Component sizes</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="do">## ```````````````          </span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="do">## size: 48 1</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="do">##    #:  1 2</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="do">## Global network properties</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="do">## `````````````````````````</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="do">## Largest connected component (LCC):</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="do">##                                  </span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="do">## Relative LCC size         0.96000</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="do">## Clustering coefficient    0.33594</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="do">## Modularity                0.53407</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="do">## Positive edge percentage 88.34951</span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="do">## Edge density              0.09131</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="do">## Natural connectivity      0.02855</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="do">## Vertex connectivity       1.00000</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="do">## Edge connectivity         1.00000</span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="do">## Average dissimilarity*    0.97035</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="do">## Average path length**     2.36912</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="do">## Whole network:</span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="do">##                                  </span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a><span class="do">## Number of components      3.00000</span></span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a><span class="do">## Clustering coefficient    0.33594</span></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a><span class="do">## Modularity                0.53407</span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a><span class="do">## Positive edge percentage 88.34951</span></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a><span class="do">## Edge density              0.08408</span></span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a><span class="do">## Natural connectivity      0.02714</span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a><span class="do">## -----</span></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="do">## *: Dissimilarity = 1 - edge weight</span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a><span class="do">## **: Path length = Units with average dissimilarity</span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a><span class="do">## Clusters</span></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a><span class="do">## - In the whole network</span></span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a><span class="do">## - Algorithm: cluster_fast_greedy</span></span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a><span class="do">## ```````````````````````````````` </span></span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a><span class="do">##                     </span></span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a><span class="do">## name: 0  1  2  3 4 5</span></span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a><span class="do">##    #: 2 12 17 10 5 4</span></span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a><span class="do">## Hubs</span></span>
<span id="cb11-45"><a href="#cb11-45" tabindex="-1"></a><span class="do">## - In alphabetical/numerical order</span></span>
<span id="cb11-46"><a href="#cb11-46" tabindex="-1"></a><span class="do">## - Based on empirical quantiles of centralities</span></span>
<span id="cb11-47"><a href="#cb11-47" tabindex="-1"></a><span class="do">## ```````````````````````````````````````````````       </span></span>
<span id="cb11-48"><a href="#cb11-48" tabindex="-1"></a><span class="do">##  190597</span></span>
<span id="cb11-49"><a href="#cb11-49" tabindex="-1"></a><span class="do">##  288134</span></span>
<span id="cb11-50"><a href="#cb11-50" tabindex="-1"></a><span class="do">##  311477</span></span>
<span id="cb11-51"><a href="#cb11-51" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-52"><a href="#cb11-52" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb11-53"><a href="#cb11-53" tabindex="-1"></a><span class="do">## Centrality measures</span></span>
<span id="cb11-54"><a href="#cb11-54" tabindex="-1"></a><span class="do">## - In decreasing order</span></span>
<span id="cb11-55"><a href="#cb11-55" tabindex="-1"></a><span class="do">## - Centrality of disconnected components is zero</span></span>
<span id="cb11-56"><a href="#cb11-56" tabindex="-1"></a><span class="do">## ````````````````````````````````````````````````</span></span>
<span id="cb11-57"><a href="#cb11-57" tabindex="-1"></a><span class="do">## Degree (unnormalized):</span></span>
<span id="cb11-58"><a href="#cb11-58" tabindex="-1"></a><span class="do">##          </span></span>
<span id="cb11-59"><a href="#cb11-59" tabindex="-1"></a><span class="do">## 288134 10</span></span>
<span id="cb11-60"><a href="#cb11-60" tabindex="-1"></a><span class="do">## 190597  9</span></span>
<span id="cb11-61"><a href="#cb11-61" tabindex="-1"></a><span class="do">## 311477  9</span></span>
<span id="cb11-62"><a href="#cb11-62" tabindex="-1"></a><span class="do">## 188236  8</span></span>
<span id="cb11-63"><a href="#cb11-63" tabindex="-1"></a><span class="do">## 199487  8</span></span>
<span id="cb11-64"><a href="#cb11-64" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-65"><a href="#cb11-65" tabindex="-1"></a><span class="do">## Betweenness centrality (normalized):</span></span>
<span id="cb11-66"><a href="#cb11-66" tabindex="-1"></a><span class="do">##               </span></span>
<span id="cb11-67"><a href="#cb11-67" tabindex="-1"></a><span class="do">## 302160 0.31360</span></span>
<span id="cb11-68"><a href="#cb11-68" tabindex="-1"></a><span class="do">## 268332 0.24144</span></span>
<span id="cb11-69"><a href="#cb11-69" tabindex="-1"></a><span class="do">## 259569 0.23404</span></span>
<span id="cb11-70"><a href="#cb11-70" tabindex="-1"></a><span class="do">## 470973 0.21462</span></span>
<span id="cb11-71"><a href="#cb11-71" tabindex="-1"></a><span class="do">## 119010 0.19611</span></span>
<span id="cb11-72"><a href="#cb11-72" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-73"><a href="#cb11-73" tabindex="-1"></a><span class="do">## Closeness centrality (normalized):</span></span>
<span id="cb11-74"><a href="#cb11-74" tabindex="-1"></a><span class="do">##               </span></span>
<span id="cb11-75"><a href="#cb11-75" tabindex="-1"></a><span class="do">## 288134 0.68426</span></span>
<span id="cb11-76"><a href="#cb11-76" tabindex="-1"></a><span class="do">## 311477 0.68413</span></span>
<span id="cb11-77"><a href="#cb11-77" tabindex="-1"></a><span class="do">## 199487 0.68099</span></span>
<span id="cb11-78"><a href="#cb11-78" tabindex="-1"></a><span class="do">## 302160 0.67518</span></span>
<span id="cb11-79"><a href="#cb11-79" tabindex="-1"></a><span class="do">## 188236 0.66852</span></span>
<span id="cb11-80"><a href="#cb11-80" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb11-81"><a href="#cb11-81" tabindex="-1"></a><span class="do">## Eigenvector centrality (normalized):</span></span>
<span id="cb11-82"><a href="#cb11-82" tabindex="-1"></a><span class="do">##               </span></span>
<span id="cb11-83"><a href="#cb11-83" tabindex="-1"></a><span class="do">## 288134 1.00000</span></span>
<span id="cb11-84"><a href="#cb11-84" tabindex="-1"></a><span class="do">## 311477 0.94417</span></span>
<span id="cb11-85"><a href="#cb11-85" tabindex="-1"></a><span class="do">## 190597 0.90794</span></span>
<span id="cb11-86"><a href="#cb11-86" tabindex="-1"></a><span class="do">## 199487 0.85439</span></span>
<span id="cb11-87"><a href="#cb11-87" tabindex="-1"></a><span class="do">## 188236 0.72684</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plotting-the-gcm-heatmap-manually">Plotting the GCM heatmap manually<a class="anchor" aria-label="anchor" href="#plotting-the-gcm-heatmap-manually"></a></h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plotHeat.html">plotHeat</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">props_spring</span><span class="op">$</span><span class="va">graphletLCC</span><span class="op">$</span><span class="va">gcm1</span>,</span>
<span>         pmat <span class="op">=</span> <span class="va">props_spring</span><span class="op">$</span><span class="va">graphletLCC</span><span class="op">$</span><span class="va">pAdjust1</span>,</span>
<span>         type <span class="op">=</span> <span class="st">"mixed"</span>,</span>
<span>         title <span class="op">=</span> <span class="st">"GCM"</span>, </span>
<span>         colorLim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>         mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add rectangles highlighting the four types of orbits</span></span>
<span><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/rect.html" class="external-link">rect</a></span><span class="op">(</span>xleft   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">0.5</span>,  <span class="fl">1.5</span>, <span class="fl">4.5</span>,  <span class="fl">7.5</span><span class="op">)</span>,</span>
<span>               ybottom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">11.5</span>,  <span class="fl">7.5</span>, <span class="fl">4.5</span>,  <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>               xright  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">1.5</span>,  <span class="fl">4.5</span>, <span class="fl">7.5</span>, <span class="fl">11.5</span><span class="op">)</span>,</span>
<span>               ytop    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10.5</span>, <span class="fl">10.5</span>, <span class="fl">7.5</span>,  <span class="fl">4.5</span><span class="op">)</span>,</span>
<span>               lwd <span class="op">=</span> <span class="fl">2</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="fl">6</span>, <span class="op">-</span><span class="fl">0.2</span>, xpd <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>     <span class="st">"Significance codes:  ***: 0.001;  **: 0.01;  *: 0.05"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_spring_heat-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="visualizing-the-network">Visualizing the network<a class="anchor" aria-label="anchor" href="#visualizing-the-network"></a></h4>
<p>We use the determined clusters as node colors and scale the node sizes according to the node’s eigenvector centrality.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># help page</span></span>
<span><span class="op">?</span><span class="va">plot.microNetProps</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_spring</span>, </span>
<span>          nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>          nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>          title1 <span class="op">=</span> <span class="st">"Network on OTU level with SPRING associations"</span>, </span>
<span>          showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated association:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_spring_3-1.png"><!-- --></p>
<p>Note that edge weights are (non-negative) similarities, however, the edges belonging to negative estimated associations are colored in red by default (<code>negDiffCol = TRUE</code>).</p>
<p>By default, a different transparency value is added to edges with an absolute weight below and above the <code>cut</code> value (arguments <code>edgeTranspLow</code> and <code>edgeTranspHigh</code>). The determined <code>cut</code> value can be read out as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">q1</span><span class="op">$</span><span class="va">Arguments</span><span class="op">$</span><span class="va">cut</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="do">##      75% </span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="do">## 0.337099</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="export-to-gephi">Export to Gephi<a class="anchor" aria-label="anchor" href="#export-to-gephi"></a></h3>
<p>Some users may be interested in how to export the network to Gephi. Here’s an example:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For Gephi, we have to generate an edge list with IDs.</span></span>
<span><span class="co"># The corresponding labels (and also further node features) are stored as node list.</span></span>
<span></span>
<span><span class="co"># Create edge object from the edge list exported by netConstruct()</span></span>
<span><span class="va">edges</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">net_spring</span><span class="op">$</span><span class="va">edgelist1</span>, <span class="va">v1</span>, <span class="va">v2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add Source and Target variables (as IDs)</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Source</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">edges</span><span class="op">$</span><span class="va">v1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Target</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">edges</span><span class="op">$</span><span class="va">v2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Type</span> <span class="op">&lt;-</span> <span class="st">"Undirected"</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Weight</span> <span class="op">&lt;-</span> <span class="va">net_spring</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">$</span><span class="va">adja</span></span>
<span></span>
<span><span class="va">nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">edges</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'v1'</span>,<span class="st">'Source'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Label"</span>, <span class="st">"Id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add category with clusters (can be used as node colors in Gephi)</span></span>
<span><span class="va">nodes</span><span class="op">$</span><span class="va">Category</span> <span class="op">&lt;-</span> <span class="va">props_spring</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clust1</span><span class="op">[</span><span class="va">nodes</span><span class="op">$</span><span class="va">Label</span><span class="op">]</span></span>
<span></span>
<span><span class="va">edges</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">edges</span>, <span class="va">Source</span>, <span class="va">Target</span>, <span class="va">Type</span>, <span class="va">Weight</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">nodes</span>, file <span class="op">=</span> <span class="st">"nodes.csv"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">edges</span>, file <span class="op">=</span> <span class="st">"edges.csv"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The exported .csv files can then be imported into Gephi.</p>
<hr></div>
<div class="section level3">
<h3 id="network-with-pearson-correlation-as-association-measure">Network with Pearson correlation as association measure<a class="anchor" aria-label="anchor" href="#network-with-pearson-correlation-as-association-measure"></a></h3>
<p>Let’s construct another network using Pearson’s correlation coefficient as association measure. The input is now a <code>phyloseq</code> object.</p>
<p>Since Pearson correlations may lead to compositional effects when applied to sequencing data, we use the clr transformation as normalization method. Zero treatment is necessary in this case.</p>
<p>A threshold of 0.3 is used as sparsification method, so that only OTUs with an absolute correlation greater than or equal to 0.3 are connected.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_pears</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>,  </span>
<span>                          measure <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>                          normMethod <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                          zeroMethod <span class="op">=</span> <span class="st">"multRepl"</span>,</span>
<span>                          sparsMethod <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>                          thresh <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="do">## Checking input arguments ... Done.</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="do">## 2 rows with zero sum removed.</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="do">## 138 taxa and 294 samples remaining.</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="do">## Zero treatment:</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="do">## Execute multRepl() ... Done.</span></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="do">## Normalization:</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="do">## Execute clr(){SpiecEasi} ... Done.</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="do">## Calculate 'pearson' associations ... Done.</span></span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a><span class="do">## Sparsify associations via 'threshold' ... Done.</span></span></code></pre></div>
<p>Network analysis and plotting:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_pears</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_pears</span>, </span>
<span>                          clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_2-1.png"><!-- --></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on OTU level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>, </span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_3-1.png"><!-- --></p>
<p>Let’s improve the visualization by changing the following arguments:</p>
<ul><li>
<code>repulsion = 0.8</code>: Place the nodes further apart.</li>
<li>
<code>rmSingles = TRUE</code>: Single nodes are removed.</li>
<li>
<code>labelScale = FALSE</code> and <code>cexLabels = 1.6</code>: All labels have equal size and are enlarged to improve readability of small node’s labels.</li>
<li>
<code>nodeSizeSpread = 3</code> (default is 4): Node sizes are more similar if the value is decreased. This argument (in combination with <code>cexNodes</code>) is useful to enlarge small nodes while keeping the size of big nodes.</li>
<li>
<code>hubBorderCol = "darkgray"</code>: Change border color for a better readability of the node labels.</li>
</ul><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.6</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on OTU level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_4-1.png"><!-- --></p>
<div class="section level4">
<h4 id="edge-filtering">Edge filtering<a class="anchor" aria-label="anchor" href="#edge-filtering"></a></h4>
<p>The network can be sparsified further using the arguments <code>edgeFilter</code> (edges are filtered before the layout is computed) and <code>edgeInvisFilter</code> (edges are removed after the layout is computed and thus just made “invisible”).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears</span>,</span>
<span>     edgeInvisFilter <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>     edgeInvisPar <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.6</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Network on OTU level with Pearson correlations"</span>,</span>
<span>                     <span class="st">"\n(edge filter: threshold = 0.4)"</span><span class="op">)</span>,</span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_5-1.png"><!-- --></p>
<hr></div>
</div>
<div class="section level3">
<h3 id="using-the-unsigned-transformation">Using the “unsigned” transformation<a class="anchor" aria-label="anchor" href="#using-the-unsigned-transformation"></a></h3>
<p>In the above network, the “signed” transformation was used to transform the estimated associations into dissimilarities. This leads to a network where strongly positive correlated taxa have a high edge weight (1 if the correlation equals 1) and strongly negative correlated taxa have a low edge weight (0 if the correlation equals -1).</p>
<p>We now use the “unsigned” transformation so that the edge weight between strongly correlated taxa is high, no matter of the sign. Hence, a correlation of -1 and 1 would lead to an edge weight of 1.</p>
<div class="section level4">
<h4 id="network-construction">Network construction<a class="anchor" aria-label="anchor" href="#network-construction"></a></h4>
<p>We can pass the network object from before to <code><a href="reference/netConstruct.html">netConstruct()</a></code> to save runtime.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_pears_unsigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">net_pears</span><span class="op">$</span><span class="va">assoEst1</span>,</span>
<span>                                   dataType <span class="op">=</span> <span class="st">"correlation"</span>, </span>
<span>                                   sparsMethod <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>                                   thresh <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                                   dissFunc <span class="op">=</span> <span class="st">"unsigned"</span>,</span>
<span>                                   verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="do">## Checking input arguments ... Done.</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="do">## Sparsify associations via 'threshold' ... Done.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="estimated-correlations-and-adjacency-values">Estimated correlations and adjacency values<a class="anchor" aria-label="anchor" href="#estimated-correlations-and-adjacency-values"></a></h4>
<p>The following histograms demonstrate how the estimated correlations are transformed into adjacencies (= sparsified similarities for weighted networks).</p>
<p>Sparsified estimated correlations:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">net_pears</span><span class="op">$</span><span class="va">assoMat1</span>, <span class="fl">100</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Estimated correlation"</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Estimated correlations after sparsification"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_hist_1-1.png"><!-- --></p>
<p>Adjacency values computed using the “signed” transformation (values different from 0 and 1 will be edges in the network):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">net_pears</span><span class="op">$</span><span class="va">adjaMat1</span>, <span class="fl">100</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Adjacency values"</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Adjacencies (with \"signed\" transformation)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_hist_2-1.png"><!-- --></p>
<p>Adjacency values computed using the “unsigned” transformation:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">net_pears_unsigned</span><span class="op">$</span><span class="va">adjaMat1</span>, <span class="fl">100</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Adjacency values"</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Adjacencies (with \"unsigned\" transformation)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_hist_3-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="network-analysis-and-plotting">Network analysis and plotting<a class="anchor" aria-label="anchor" href="#network-analysis-and-plotting"></a></h4>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_pears_unsigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_pears_unsigned</span>, </span>
<span>                                   clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                                   gcmHeat <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears_unsigned</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.6</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network with Pearson correlations and \"unsigned\" transformation"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_pears_unsigned_4-1.png"><!-- --></p>
<p>While with the “signed” transformation, positive correlated taxa are likely to belong to the same cluster, with the “unsigned” transformation clusters contain strongly positive and negative correlated taxa.</p>
<hr></div>
</div>
<div class="section level3">
<h3 id="network-on-genus-level">Network on genus level<a class="anchor" aria-label="anchor" href="#network-on-genus-level"></a></h3>
<p>We now construct a further network, where OTUs are agglomerated to genera.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"amgut2.filt.phy"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Agglomerate to genus level</span></span>
<span><span class="va">amgut_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_glom.html" class="external-link">tax_glom</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>, taxrank <span class="op">=</span> <span class="st">"Rank6"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Taxonomic table</span></span>
<span><span class="va">taxtab</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename taxonomic table and make Rank6 (genus) unique</span></span>
<span><span class="va">amgut_genus_renamed</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/renameTaxa.html">renameTaxa</a></span><span class="op">(</span><span class="va">amgut_genus</span>, </span>
<span>                                  pat <span class="op">=</span> <span class="st">"&lt;name&gt;"</span>, </span>
<span>                                  substPat <span class="op">=</span> <span class="st">"&lt;name&gt;_&lt;subst_name&gt;(&lt;subst_R&gt;)"</span>,</span>
<span>                                  numDupli <span class="op">=</span> <span class="st">"Rank6"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="do">## Column 7 contains NAs only and is ignored.</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Network construction and analysis</span></span>
<span><span class="va">net_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut_genus_renamed</span>,</span>
<span>                          taxRank <span class="op">=</span> <span class="st">"Rank6"</span>,</span>
<span>                          measure <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>                          zeroMethod <span class="op">=</span> <span class="st">"multRepl"</span>,</span>
<span>                          normMethod <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                          sparsMethod <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>                          thresh <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="do">## Checking input arguments ...</span></span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a><span class="do">## Done.</span></span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a><span class="do">## 2 rows with zero sum removed.</span></span>
<span id="cb34-6"><a href="#cb34-6" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" tabindex="-1"></a><span class="do">## 43 taxa and 294 samples remaining.</span></span>
<span id="cb34-8"><a href="#cb34-8" tabindex="-1"></a></span>
<span id="cb34-9"><a href="#cb34-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb34-10"><a href="#cb34-10" tabindex="-1"></a><span class="do">## Zero treatment:</span></span>
<span id="cb34-11"><a href="#cb34-11" tabindex="-1"></a></span>
<span id="cb34-12"><a href="#cb34-12" tabindex="-1"></a><span class="do">## Execute multRepl() ... Done.</span></span>
<span id="cb34-13"><a href="#cb34-13" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb34-14"><a href="#cb34-14" tabindex="-1"></a><span class="do">## Normalization:</span></span>
<span id="cb34-15"><a href="#cb34-15" tabindex="-1"></a><span class="do">## Execute clr(){SpiecEasi} ... Done.</span></span>
<span id="cb34-16"><a href="#cb34-16" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb34-17"><a href="#cb34-17" tabindex="-1"></a><span class="do">## Calculate 'pearson' associations ... Done.</span></span>
<span id="cb34-18"><a href="#cb34-18" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb34-19"><a href="#cb34-19" tabindex="-1"></a><span class="do">## Sparsify associations via 'threshold' ... Done.</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_genus</span>, clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_genus_1-1.png"><!-- --></p>
<div class="section level4">
<h4 id="network-plots">Network plots<a class="anchor" aria-label="anchor" href="#network-plots"></a></h4>
<p>Modifications:</p>
<ul><li>Fruchterman-Reingold layout algorithm from <code>igraph</code> package used (passed to <code>plot</code> as matrix)</li>
<li>Shortened labels (using the “intelligent” method, which avoids duplicates)</li>
<li>Fixed node sizes, where hubs are enlarged</li>
<li>Node color is gray for all nodes (transparancy is lower for hub nodes by default)</li>
</ul><div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute layout</span></span>
<span><span class="va">graph3</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_adjacency_matrix.html" class="external-link">graph_from_adjacency_matrix</a></span><span class="op">(</span><span class="va">net_genus</span><span class="op">$</span><span class="va">adjaMat1</span>, </span>
<span>                                              weighted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span><span class="va">lay_fr</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/layout_with_fr.html" class="external-link">layout_with_fr</a></span><span class="op">(</span><span class="va">graph3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Row names of the layout matrix must match the node names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lay_fr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">net_genus</span><span class="op">$</span><span class="va">adjaMat1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_genus</span>,</span>
<span>     layout <span class="op">=</span> <span class="va">lay_fr</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"intelligent"</span>,</span>
<span>     labelLength <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     labelPattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"'"</span>, <span class="fl">3</span>, <span class="st">"'"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"fix"</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"gray"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     cexHubs <span class="op">=</span> <span class="fl">1.1</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on genus level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_genus_2-1.png"><!-- --></p>
<p>Since the above visualization is obviously not optimal, we make further adjustments:</p>
<ul><li>This time, the Fruchterman-Reingold layout algorithm is computed within the plot function and thus applied to the “reduced” network without singletons</li>
<li>Labels are not scaled to node sizes</li>
<li>Single nodes are removed</li>
<li>Node sizes are scaled to the column sums of clr-transformed data</li>
<li>Node colors represent the determined clusters</li>
<li>Border color of hub nodes is changed from black to darkgray</li>
<li>Label size of hubs is enlarged</li>
</ul><div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_genus</span>,</span>
<span>     layout <span class="op">=</span> <span class="st">"layout_with_fr"</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"intelligent"</span>,</span>
<span>     labelLength <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     labelPattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"'"</span>, <span class="fl">3</span>, <span class="st">"'"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on genus level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_genus_3-1.png"><!-- --></p>
<p>Let’s check whether the largest nodes are actually those with highest column sums in the matrix with normalized counts returned by <code><a href="reference/netConstruct.html">netConstruct()</a></code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">net_genus</span><span class="op">$</span><span class="va">normCounts1</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="do">##             Bacteroides              Klebsiella        Faecalibacterium </span></span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a><span class="do">##               1200.7971               1137.4928                708.0877 </span></span>
<span id="cb39-3"><a href="#cb39-3" tabindex="-1"></a><span class="do">##      5_Clostridiales(O)    2_Ruminococcaceae(F)    3_Lachnospiraceae(F) </span></span>
<span id="cb39-4"><a href="#cb39-4" tabindex="-1"></a><span class="do">##                549.2647                502.1889                493.7558 </span></span>
<span id="cb39-5"><a href="#cb39-5" tabindex="-1"></a><span class="do">## 6_Enterobacteriaceae(F)               Roseburia         Parabacteroides </span></span>
<span id="cb39-6"><a href="#cb39-6" tabindex="-1"></a><span class="do">##                363.3841                333.8737                328.0495 </span></span>
<span id="cb39-7"><a href="#cb39-7" tabindex="-1"></a><span class="do">##             Coprococcus </span></span>
<span id="cb39-8"><a href="#cb39-8" tabindex="-1"></a><span class="do">##                274.4082</span></span></code></pre></div>
<p>In order to further improve our plot, we use the following modifications:</p>
<ul><li>This time, we choose the “spring” layout as part of <code>qgraph()</code> (the function is generally used for network plotting in NetCoMi)</li>
<li>A repulsion value below 1 places the nodes further apart</li>
<li>Labels are not shortened anymore</li>
<li>Nodes (bacteria on genus level) are colored according to the respective phylum</li>
<li>Edges representing positive associations are colored in blue, negative ones in orange (just to give an example for alternative edge coloring)</li>
<li>Transparency is increased for edges with high weight to improve the readability of node labels</li>
</ul><div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get phyla names</span></span>
<span><span class="va">taxtab</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus_renamed</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"p__"</span>, <span class="st">""</span>, <span class="va">taxtab</span><span class="op">[</span>, <span class="st">"Rank2"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">taxtab</span><span class="op">[</span>, <span class="st">"Rank6"</span><span class="op">]</span></span>
<span><span class="co">#table(phyla)</span></span>
<span></span>
<span><span class="co"># Define phylum colors</span></span>
<span><span class="va">phylcol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cyan"</span>, <span class="st">"blue3"</span>, <span class="st">"red"</span>, <span class="st">"lawngreen"</span>, <span class="st">"yellow"</span>, <span class="st">"deeppink"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_genus</span>,</span>
<span>     layout <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.84</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     charToRm <span class="op">=</span> <span class="st">"g__"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">4</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"feature"</span>, </span>
<span>     featVecCol <span class="op">=</span> <span class="va">phyla</span>, </span>
<span>     colorVec <span class="op">=</span>  <span class="va">phylcol</span>,</span>
<span>     posCol <span class="op">=</span> <span class="st">"darkturquoise"</span>, </span>
<span>     negCol <span class="op">=</span> <span class="st">"orange"</span>,</span>
<span>     edgeTranspLow <span class="op">=</span> <span class="fl">0</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">40</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on genus level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span><span class="va">phylcol_transp</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/colToTransp.html">colToTransp</a></span><span class="op">(</span><span class="va">phylcol</span>, <span class="fl">60</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.2</span>, <span class="fl">1.2</span>, cex <span class="op">=</span> <span class="fl">2</span>, pt.cex <span class="op">=</span> <span class="fl">2.5</span>, title <span class="op">=</span> <span class="st">"Phylum:"</span>, </span>
<span>       legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">phylcol_transp</span>, bty <span class="op">=</span> <span class="st">"n"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkturquoise"</span>,<span class="st">"orange"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/single_genus_5-1.png"><!-- --></p>
<hr></div>
</div>
<div class="section level3">
<h3 id="using-an-association-matrix-as-input">Using an association matrix as input<a class="anchor" aria-label="anchor" href="#using-an-association-matrix-as-input"></a></h3>
<p>The QMP data set provided by the <code>SPRING</code> package is used to demonstrate how NetCoMi is used to analyze a precomputed network (given as association matrix).</p>
<p>The data set contains quantitative count data (true absolute values), which SPRING can deal with. See <code><a href="https://rdrr.io/pkg/SPRING/man/QMP.html" class="external-link">?QMP</a></code> for details.</p>
<p><code>nlambda</code> and <code>rep.num</code> are set to 10 for a decreased execution time, but should be higher for real data.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SPRING</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the QMP data set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"QMP"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Run SPRING for association estimation</span></span>
<span><span class="va">fit_spring</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SPRING/man/SPRING.html" class="external-link">SPRING</a></span><span class="op">(</span><span class="va">QMP</span>, </span>
<span>                     quantitative <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     lambdaseq <span class="op">=</span> <span class="st">"data-specific"</span>,</span>
<span>                     nlambda <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                     rep.num <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                     seed <span class="op">=</span> <span class="fl">123456</span>, </span>
<span>                     ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     Rmethod <span class="op">=</span> <span class="st">"approx"</span>,</span>
<span>                     verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Optimal lambda</span></span>
<span><span class="va">opt.K</span> <span class="op">&lt;-</span> <span class="va">fit_spring</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">stars</span><span class="op">$</span><span class="va">opt.index</span></span>
<span>    </span>
<span><span class="co"># Association matrix</span></span>
<span><span class="va">assoMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">SpiecEasi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpiecEasi/man/symBeta.html" class="external-link">symBeta</a></span><span class="op">(</span><span class="va">fit_spring</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">est</span><span class="op">$</span><span class="va">beta</span><span class="op">[[</span><span class="va">opt.K</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                        mode <span class="op">=</span> <span class="st">"ave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">assoMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assoMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">QMP</span><span class="op">)</span></span></code></pre></div>
<p>The association matrix is now passed to <code>netConstruct</code> to start the usual NetCoMi workflow. Note that the <code>dataType</code> argument must be set appropriately.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Network construction and analysis</span></span>
<span><span class="va">net_asso</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">assoMat</span>,</span>
<span>                         dataType <span class="op">=</span> <span class="st">"condDependence"</span>,</span>
<span>                         sparsMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">props_asso</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_asso</span>, clustMethod <span class="op">=</span> <span class="st">"hierarchical"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/association_input_2-1.png"><!-- --></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_asso</span>,</span>
<span>     layout <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"gray60"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">2.2</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network for QMP data"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated association:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/association_input_3-1.png"><!-- --></p>
<hr></div>
<div class="section level3">
<h3 id="network-comparison">Network comparison<a class="anchor" aria-label="anchor" href="#network-comparison"></a></h3>
<p>Now let’s look how NetCoMi is used to compare two networks.</p>
<div class="section level4">
<h4 id="network-construction-1">Network construction<a class="anchor" aria-label="anchor" href="#network-construction-1"></a></h4>
<p>The data set is split by <code>"SEASONAL_ALLERGIES"</code> leading to two subsets of samples (with and without seasonal allergies). We ignore the “None” group.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Split the phyloseq object into two groups</span></span>
<span><span class="va">amgut_season_yes</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>, </span>
<span>                                             <span class="va">SEASONAL_ALLERGIES</span> <span class="op">==</span> <span class="st">"yes"</span><span class="op">)</span></span>
<span><span class="va">amgut_season_no</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>, </span>
<span>                                            <span class="va">SEASONAL_ALLERGIES</span> <span class="op">==</span> <span class="st">"no"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">amgut_season_yes</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="do">## phyloseq-class experiment-level object</span></span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a><span class="do">## otu_table()   OTU Table:         [ 138 taxa and 121 samples ]</span></span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a><span class="do">## sample_data() Sample Data:       [ 121 samples by 166 sample variables ]</span></span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a><span class="do">## tax_table()   Taxonomy Table:    [ 138 taxa by 7 taxonomic ranks ]</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">amgut_season_no</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="do">## phyloseq-class experiment-level object</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a><span class="do">## otu_table()   OTU Table:         [ 138 taxa and 163 samples ]</span></span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a><span class="do">## sample_data() Sample Data:       [ 163 samples by 166 sample variables ]</span></span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a><span class="do">## tax_table()   Taxonomy Table:    [ 138 taxa by 7 taxonomic ranks ]</span></span></code></pre></div>
<p>The 50 nodes with highest variance are selected for network construction to get smaller networks.</p>
<p>We filter the 121 samples (sample size of the smaller group) with highest frequency to make the sample sizes equal and thus ensure comparability.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_yes</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">amgut_season_yes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Network construction</span></span>
<span><span class="va">net_season</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">amgut_season_no</span>, </span>
<span>                           data2 <span class="op">=</span> <span class="va">amgut_season_yes</span>,  </span>
<span>                           filtTax <span class="op">=</span> <span class="st">"highestVar"</span>,</span>
<span>                           filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestVar <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>                           filtSamp <span class="op">=</span> <span class="st">"highestFreq"</span>,</span>
<span>                           filtSampPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestFreq <span class="op">=</span> <span class="va">n_yes</span><span class="op">)</span>,</span>
<span>                           measure <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>                           measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nlambda <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                                             rep.num <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                             Rmethod <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span>,</span>
<span>                           normMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           zeroMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                           sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                           seed <span class="op">=</span> <span class="fl">123456</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="do">## Checking input arguments ... Done.</span></span>
<span id="cb49-2"><a href="#cb49-2" tabindex="-1"></a><span class="do">## Data filtering ...</span></span>
<span id="cb49-3"><a href="#cb49-3" tabindex="-1"></a><span class="do">## 42 samples removed in data set 1.</span></span>
<span id="cb49-4"><a href="#cb49-4" tabindex="-1"></a><span class="do">## 0 samples removed in data set 2.</span></span>
<span id="cb49-5"><a href="#cb49-5" tabindex="-1"></a><span class="do">## 96 taxa removed in each data set.</span></span>
<span id="cb49-6"><a href="#cb49-6" tabindex="-1"></a><span class="do">## 1 rows with zero sum removed in group 2.</span></span>
<span id="cb49-7"><a href="#cb49-7" tabindex="-1"></a><span class="do">## 42 taxa and 121 samples remaining in group 1.</span></span>
<span id="cb49-8"><a href="#cb49-8" tabindex="-1"></a><span class="do">## 42 taxa and 120 samples remaining in group 2.</span></span>
<span id="cb49-9"><a href="#cb49-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb49-10"><a href="#cb49-10" tabindex="-1"></a><span class="do">## Calculate 'spring' associations ... Done.</span></span>
<span id="cb49-11"><a href="#cb49-11" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb49-12"><a href="#cb49-12" tabindex="-1"></a><span class="do">## Calculate associations in group 2 ... Done.</span></span></code></pre></div>
<p>Alternatively, a group vector could be passed to <code>group</code>, according to which the data set is split into two groups:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get count table</span></span>
<span><span class="va">countMat</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html" class="external-link">otu_table</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># netConstruct() expects samples in rows</span></span>
<span><span class="va">countMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="va">countMat</span>, <span class="st">"matrix"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">group_vec</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/get_variable.html" class="external-link">get_variable</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>, <span class="st">"SEASONAL_ALLERGIES"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select the two groups of interest (level "none" is excluded)</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">group_vec</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"no"</span>, <span class="st">"yes"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">group_vec</span> <span class="op">&lt;-</span> <span class="va">group_vec</span><span class="op">[</span><span class="va">sel</span><span class="op">]</span></span>
<span><span class="va">countMat</span> <span class="op">&lt;-</span> <span class="va">countMat</span><span class="op">[</span><span class="va">sel</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">net_season</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">countMat</span>, </span>
<span>                           group <span class="op">=</span> <span class="va">group_vec</span>, </span>
<span>                           filtTax <span class="op">=</span> <span class="st">"highestVar"</span>,</span>
<span>                           filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestVar <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>                           filtSamp <span class="op">=</span> <span class="st">"highestFreq"</span>,</span>
<span>                           filtSampPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestFreq <span class="op">=</span> <span class="va">n_yes</span><span class="op">)</span>,</span>
<span>                           measure <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>                           measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nlambda<span class="op">=</span><span class="fl">10</span>, </span>
<span>                                             rep.num<span class="op">=</span><span class="fl">10</span>,</span>
<span>                                             Rmethod <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span>,</span>
<span>                           normMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           zeroMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                           sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                           seed <span class="op">=</span> <span class="fl">123456</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="network-analysis">Network analysis<a class="anchor" aria-label="anchor" href="#network-analysis"></a></h4>
<p>The object returned by <code><a href="reference/netConstruct.html">netConstruct()</a></code> containing both networks is again passed to <code><a href="reference/netAnalyze.html">netAnalyze()</a></code>. Network properties are computed for both networks simultaneously.</p>
<p>To demonstrate further functionalities of <code><a href="reference/netAnalyze.html">netAnalyze()</a></code>, we play around with the available arguments, even if the chosen setting might not be optimal.</p>
<ul><li>
<code>centrLCC = FALSE</code>: Centralities are calculated for all nodes (not only for the largest connected component).</li>
<li>
<code>avDissIgnoreInf = TRUE</code>: Nodes with an infinite dissimilarity are ignored when calculating the average dissimilarity.</li>
<li>
<code>sPathNorm = FALSE</code>: Shortest paths are not normalized by average dissimilarity.</li>
<li>
<code>hubPar = c("degree", "eigenvector")</code>: Hubs are nodes with highest degree and eigenvector centrality at the same time.</li>
<li>
<code>lnormFit = TRUE</code> and <code>hubQuant = 0.9</code>: A log-normal distribution is fitted to the centrality values to identify nodes with “highest” centrality values. Here, a node is identified as hub if for each of the three centrality measures, the node’s centrality value is above the 90% quantile of the fitted log-normal distribution.</li>
<li>The non-normalized centralities are used for all four measures.</li>
</ul><p><strong>Note! The arguments must be set carefully, depending on the research questions. NetCoMi’s default values are not generally preferable in all practical cases!</strong></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_season</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_season</span>, </span>
<span>                           centrLCC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           avDissIgnoreInf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           sPathNorm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                           hubPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"degree"</span>, <span class="st">"eigenvector"</span><span class="op">)</span>,</span>
<span>                           hubQuant <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>                           lnormFit <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           normDeg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           normBetw <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           normClose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                           normEigen <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/netcomp_spring_3-1.png"><!-- --></p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">props_season</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a><span class="do">## Component sizes</span></span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a><span class="do">## ```````````````</span></span>
<span id="cb53-4"><a href="#cb53-4" tabindex="-1"></a><span class="do">## group '1':           </span></span>
<span id="cb53-5"><a href="#cb53-5" tabindex="-1"></a><span class="do">## size: 28  1</span></span>
<span id="cb53-6"><a href="#cb53-6" tabindex="-1"></a><span class="do">##    #:  1 14</span></span>
<span id="cb53-7"><a href="#cb53-7" tabindex="-1"></a><span class="do">## group '2':            </span></span>
<span id="cb53-8"><a href="#cb53-8" tabindex="-1"></a><span class="do">## size: 31 8 1</span></span>
<span id="cb53-9"><a href="#cb53-9" tabindex="-1"></a><span class="do">##    #:  1 1 3</span></span>
<span id="cb53-10"><a href="#cb53-10" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb53-11"><a href="#cb53-11" tabindex="-1"></a><span class="do">## Global network properties</span></span>
<span id="cb53-12"><a href="#cb53-12" tabindex="-1"></a><span class="do">## `````````````````````````</span></span>
<span id="cb53-13"><a href="#cb53-13" tabindex="-1"></a><span class="do">## Largest connected component (LCC):</span></span>
<span id="cb53-14"><a href="#cb53-14" tabindex="-1"></a><span class="do">##                          group '1' group '2'</span></span>
<span id="cb53-15"><a href="#cb53-15" tabindex="-1"></a><span class="do">## Relative LCC size          0.66667   0.73810</span></span>
<span id="cb53-16"><a href="#cb53-16" tabindex="-1"></a><span class="do">## Clustering coefficient     0.15161   0.27111</span></span>
<span id="cb53-17"><a href="#cb53-17" tabindex="-1"></a><span class="do">## Modularity                 0.62611   0.45823</span></span>
<span id="cb53-18"><a href="#cb53-18" tabindex="-1"></a><span class="do">## Positive edge percentage  86.66667 100.00000</span></span>
<span id="cb53-19"><a href="#cb53-19" tabindex="-1"></a><span class="do">## Edge density               0.07937   0.12473</span></span>
<span id="cb53-20"><a href="#cb53-20" tabindex="-1"></a><span class="do">## Natural connectivity       0.04539   0.04362</span></span>
<span id="cb53-21"><a href="#cb53-21" tabindex="-1"></a><span class="do">## Vertex connectivity        1.00000   1.00000</span></span>
<span id="cb53-22"><a href="#cb53-22" tabindex="-1"></a><span class="do">## Edge connectivity          1.00000   1.00000</span></span>
<span id="cb53-23"><a href="#cb53-23" tabindex="-1"></a><span class="do">## Average dissimilarity*     0.67251   0.68178</span></span>
<span id="cb53-24"><a href="#cb53-24" tabindex="-1"></a><span class="do">## Average path length**      3.40008   1.86767</span></span>
<span id="cb53-25"><a href="#cb53-25" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-26"><a href="#cb53-26" tabindex="-1"></a><span class="do">## Whole network:</span></span>
<span id="cb53-27"><a href="#cb53-27" tabindex="-1"></a><span class="do">##                          group '1' group '2'</span></span>
<span id="cb53-28"><a href="#cb53-28" tabindex="-1"></a><span class="do">## Number of components      15.00000   5.00000</span></span>
<span id="cb53-29"><a href="#cb53-29" tabindex="-1"></a><span class="do">## Clustering coefficient     0.15161   0.29755</span></span>
<span id="cb53-30"><a href="#cb53-30" tabindex="-1"></a><span class="do">## Modularity                 0.62611   0.55684</span></span>
<span id="cb53-31"><a href="#cb53-31" tabindex="-1"></a><span class="do">## Positive edge percentage  86.66667 100.00000</span></span>
<span id="cb53-32"><a href="#cb53-32" tabindex="-1"></a><span class="do">## Edge density               0.03484   0.08130</span></span>
<span id="cb53-33"><a href="#cb53-33" tabindex="-1"></a><span class="do">## Natural connectivity       0.02826   0.03111</span></span>
<span id="cb53-34"><a href="#cb53-34" tabindex="-1"></a><span class="do">## -----</span></span>
<span id="cb53-35"><a href="#cb53-35" tabindex="-1"></a><span class="do">## *: Dissimilarity = 1 - edge weight</span></span>
<span id="cb53-36"><a href="#cb53-36" tabindex="-1"></a><span class="do">## **: Path length = Sum of dissimilarities along the path</span></span>
<span id="cb53-37"><a href="#cb53-37" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-38"><a href="#cb53-38" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb53-39"><a href="#cb53-39" tabindex="-1"></a><span class="do">## Clusters</span></span>
<span id="cb53-40"><a href="#cb53-40" tabindex="-1"></a><span class="do">## - In the whole network</span></span>
<span id="cb53-41"><a href="#cb53-41" tabindex="-1"></a><span class="do">## - Algorithm: cluster_fast_greedy</span></span>
<span id="cb53-42"><a href="#cb53-42" tabindex="-1"></a><span class="do">## ```````````````````````````````` </span></span>
<span id="cb53-43"><a href="#cb53-43" tabindex="-1"></a><span class="do">## group '1':                  </span></span>
<span id="cb53-44"><a href="#cb53-44" tabindex="-1"></a><span class="do">## name:  0 1 2 3 4 5</span></span>
<span id="cb53-45"><a href="#cb53-45" tabindex="-1"></a><span class="do">##    #: 14 7 6 5 4 6</span></span>
<span id="cb53-46"><a href="#cb53-46" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-47"><a href="#cb53-47" tabindex="-1"></a><span class="do">## group '2':                  </span></span>
<span id="cb53-48"><a href="#cb53-48" tabindex="-1"></a><span class="do">## name: 0 1  2 3 4 5</span></span>
<span id="cb53-49"><a href="#cb53-49" tabindex="-1"></a><span class="do">##    #: 3 5 14 4 8 8</span></span>
<span id="cb53-50"><a href="#cb53-50" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-51"><a href="#cb53-51" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb53-52"><a href="#cb53-52" tabindex="-1"></a><span class="do">## Hubs</span></span>
<span id="cb53-53"><a href="#cb53-53" tabindex="-1"></a><span class="do">## - In alphabetical/numerical order</span></span>
<span id="cb53-54"><a href="#cb53-54" tabindex="-1"></a><span class="do">## - Based on log-normal quantiles of centralities</span></span>
<span id="cb53-55"><a href="#cb53-55" tabindex="-1"></a><span class="do">## ```````````````````````````````````````````````</span></span>
<span id="cb53-56"><a href="#cb53-56" tabindex="-1"></a><span class="do">##  group '1' group '2'</span></span>
<span id="cb53-57"><a href="#cb53-57" tabindex="-1"></a><span class="do">##     307981    322235</span></span>
<span id="cb53-58"><a href="#cb53-58" tabindex="-1"></a><span class="do">##               363302</span></span>
<span id="cb53-59"><a href="#cb53-59" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-60"><a href="#cb53-60" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb53-61"><a href="#cb53-61" tabindex="-1"></a><span class="do">## Centrality measures</span></span>
<span id="cb53-62"><a href="#cb53-62" tabindex="-1"></a><span class="do">## - In decreasing order</span></span>
<span id="cb53-63"><a href="#cb53-63" tabindex="-1"></a><span class="do">## - Computed for the complete network</span></span>
<span id="cb53-64"><a href="#cb53-64" tabindex="-1"></a><span class="do">## ````````````````````````````````````</span></span>
<span id="cb53-65"><a href="#cb53-65" tabindex="-1"></a><span class="do">## Degree (unnormalized):</span></span>
<span id="cb53-66"><a href="#cb53-66" tabindex="-1"></a><span class="do">##         group '1' group '2'</span></span>
<span id="cb53-67"><a href="#cb53-67" tabindex="-1"></a><span class="do">##  307981         5         2</span></span>
<span id="cb53-68"><a href="#cb53-68" tabindex="-1"></a><span class="do">##    9715         5         5</span></span>
<span id="cb53-69"><a href="#cb53-69" tabindex="-1"></a><span class="do">##  364563         4         4</span></span>
<span id="cb53-70"><a href="#cb53-70" tabindex="-1"></a><span class="do">##  259569         4         5</span></span>
<span id="cb53-71"><a href="#cb53-71" tabindex="-1"></a><span class="do">##  322235         3         9</span></span>
<span id="cb53-72"><a href="#cb53-72" tabindex="-1"></a><span class="do">##            ______    ______</span></span>
<span id="cb53-73"><a href="#cb53-73" tabindex="-1"></a><span class="do">##  322235         3         9</span></span>
<span id="cb53-74"><a href="#cb53-74" tabindex="-1"></a><span class="do">##  363302         3         9</span></span>
<span id="cb53-75"><a href="#cb53-75" tabindex="-1"></a><span class="do">##  158660         2         6</span></span>
<span id="cb53-76"><a href="#cb53-76" tabindex="-1"></a><span class="do">##  188236         3         5</span></span>
<span id="cb53-77"><a href="#cb53-77" tabindex="-1"></a><span class="do">##  259569         4         5</span></span>
<span id="cb53-78"><a href="#cb53-78" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-79"><a href="#cb53-79" tabindex="-1"></a><span class="do">## Betweenness centrality (unnormalized):</span></span>
<span id="cb53-80"><a href="#cb53-80" tabindex="-1"></a><span class="do">##         group '1' group '2'</span></span>
<span id="cb53-81"><a href="#cb53-81" tabindex="-1"></a><span class="do">##  307981       231         0</span></span>
<span id="cb53-82"><a href="#cb53-82" tabindex="-1"></a><span class="do">##  331820       170         9</span></span>
<span id="cb53-83"><a href="#cb53-83" tabindex="-1"></a><span class="do">##  158660       162        80</span></span>
<span id="cb53-84"><a href="#cb53-84" tabindex="-1"></a><span class="do">##  188236       161        85</span></span>
<span id="cb53-85"><a href="#cb53-85" tabindex="-1"></a><span class="do">##  322235       159       126</span></span>
<span id="cb53-86"><a href="#cb53-86" tabindex="-1"></a><span class="do">##            ______    ______</span></span>
<span id="cb53-87"><a href="#cb53-87" tabindex="-1"></a><span class="do">##  322235       159       126</span></span>
<span id="cb53-88"><a href="#cb53-88" tabindex="-1"></a><span class="do">##  363302        74        93</span></span>
<span id="cb53-89"><a href="#cb53-89" tabindex="-1"></a><span class="do">##  188236       161        85</span></span>
<span id="cb53-90"><a href="#cb53-90" tabindex="-1"></a><span class="do">##  158660       162        80</span></span>
<span id="cb53-91"><a href="#cb53-91" tabindex="-1"></a><span class="do">##  326792        17        58</span></span>
<span id="cb53-92"><a href="#cb53-92" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-93"><a href="#cb53-93" tabindex="-1"></a><span class="do">## Closeness centrality (unnormalized):</span></span>
<span id="cb53-94"><a href="#cb53-94" tabindex="-1"></a><span class="do">##         group '1' group '2'</span></span>
<span id="cb53-95"><a href="#cb53-95" tabindex="-1"></a><span class="do">##  307981  18.17276   7.80251</span></span>
<span id="cb53-96"><a href="#cb53-96" tabindex="-1"></a><span class="do">##    9715   15.8134   9.27254</span></span>
<span id="cb53-97"><a href="#cb53-97" tabindex="-1"></a><span class="do">##  188236   15.7949  23.24055</span></span>
<span id="cb53-98"><a href="#cb53-98" tabindex="-1"></a><span class="do">##  301645  15.30177   9.01509</span></span>
<span id="cb53-99"><a href="#cb53-99" tabindex="-1"></a><span class="do">##  364563  14.73566  21.21352</span></span>
<span id="cb53-100"><a href="#cb53-100" tabindex="-1"></a><span class="do">##            ______    ______</span></span>
<span id="cb53-101"><a href="#cb53-101" tabindex="-1"></a><span class="do">##  322235  13.50232  26.36749</span></span>
<span id="cb53-102"><a href="#cb53-102" tabindex="-1"></a><span class="do">##  363302  12.30297  24.19703</span></span>
<span id="cb53-103"><a href="#cb53-103" tabindex="-1"></a><span class="do">##  158660  13.07106  23.31577</span></span>
<span id="cb53-104"><a href="#cb53-104" tabindex="-1"></a><span class="do">##  188236   15.7949  23.24055</span></span>
<span id="cb53-105"><a href="#cb53-105" tabindex="-1"></a><span class="do">##  326792  14.61391  22.52157</span></span>
<span id="cb53-106"><a href="#cb53-106" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb53-107"><a href="#cb53-107" tabindex="-1"></a><span class="do">## Eigenvector centrality (unnormalized):</span></span>
<span id="cb53-108"><a href="#cb53-108" tabindex="-1"></a><span class="do">##         group '1' group '2'</span></span>
<span id="cb53-109"><a href="#cb53-109" tabindex="-1"></a><span class="do">##  307981   0.53313   0.06912</span></span>
<span id="cb53-110"><a href="#cb53-110" tabindex="-1"></a><span class="do">##    9715   0.44398   0.10788</span></span>
<span id="cb53-111"><a href="#cb53-111" tabindex="-1"></a><span class="do">##  301645   0.41878   0.08572</span></span>
<span id="cb53-112"><a href="#cb53-112" tabindex="-1"></a><span class="do">##  326792   0.27033   0.15727</span></span>
<span id="cb53-113"><a href="#cb53-113" tabindex="-1"></a><span class="do">##  188236   0.25824   0.21162</span></span>
<span id="cb53-114"><a href="#cb53-114" tabindex="-1"></a><span class="do">##            ______    ______</span></span>
<span id="cb53-115"><a href="#cb53-115" tabindex="-1"></a><span class="do">##  322235   0.01749   0.29705</span></span>
<span id="cb53-116"><a href="#cb53-116" tabindex="-1"></a><span class="do">##  363302   0.03526   0.28512</span></span>
<span id="cb53-117"><a href="#cb53-117" tabindex="-1"></a><span class="do">##  188236   0.25824   0.21162</span></span>
<span id="cb53-118"><a href="#cb53-118" tabindex="-1"></a><span class="do">##  194648   0.00366   0.19448</span></span>
<span id="cb53-119"><a href="#cb53-119" tabindex="-1"></a><span class="do">##  184983    0.0917    0.1854</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visual-network-comparison">Visual network comparison<a class="anchor" aria-label="anchor" href="#visual-network-comparison"></a></h4>
<p>First, the layout is computed separately in both groups (qgraph’s “spring” layout in this case).</p>
<p>Node sizes are scaled according to the mclr-transformed data since <code>SPRING</code> uses the mclr transformation as normalization method.</p>
<p>Node colors represent clusters. Note that by default, two clusters have the same color in both groups if they have at least two nodes in common (<code>sameColThresh = 2</code>). Set <code>sameClustCol</code> to <code>FALSE</code> to get different cluster colors.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_season</span>, </span>
<span>     sameLayout <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"mclr"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.5</span>, </span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">3.7</span>,</span>
<span>     groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No seasonal allergies"</span>, <span class="st">"Seasonal allergies"</span><span class="op">)</span>,</span>
<span>     hubBorderCol  <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottom"</span>, title <span class="op">=</span> <span class="st">"estimated association:"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, </span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, inset <span class="op">=</span> <span class="fl">0.02</span>, cex <span class="op">=</span> <span class="fl">4</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">4</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/netcomp_spring_4-1.png"><!-- --></p>
<p>Using different layouts leads to a “nice-looking” network plot for each group, however, it is difficult to identify group differences at first glance.</p>
<p>Thus, we now use the same layout in both groups. In the following, the layout is computed for group 1 (the left network) and taken over for group 2.</p>
<p><code>rmSingles</code> is set to <code>"inboth"</code> because only nodes that are unconnected in both groups can be removed if the same layout is used.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_season</span>, </span>
<span>     sameLayout <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>     layoutGroup <span class="op">=</span> <span class="fl">1</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="st">"inboth"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"mclr"</span>, </span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.5</span>, </span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">3.8</span>,</span>
<span>     groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No seasonal allergies"</span>, <span class="st">"Seasonal allergies"</span><span class="op">)</span>,</span>
<span>     hubBorderCol  <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottom"</span>, title <span class="op">=</span> <span class="st">"estimated association:"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, </span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, inset <span class="op">=</span> <span class="fl">0.02</span>, cex <span class="op">=</span> <span class="fl">4</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">4</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/netcomp_spring_5-1.png"><!-- --></p>
<p>In the above plot, we can see clear differences between the groups. The OTU “322235”, for instance, is more strongly connected in the “Seasonal allergies” group than in the group without seasonal allergies, which is why it is a hub on the right, but not on the left.</p>
<p>However, if the layout of one group is simply taken over to the other, one of the networks (here the “seasonal allergies” group) is usually not that nice-looking due to the long edges. Therefore, NetCoMi (&gt;= 1.0.2) offers a further option (<code>layoutGroup = "union"</code>), where a union of the two layouts is used in both groups. In doing so, the nodes are placed as optimal as possible equally for both networks.</p>
<p><em>The idea and R code for this functionality were provided by <a href="https://github.com/muellsen?tab=followers" class="external-link">Christian L. Müller</a> and <a href="https://www.iq.harvard.edu/people/alice-sommer" class="external-link">Alice Sommer</a></em></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_season</span>, </span>
<span>     sameLayout <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>     layoutGroup <span class="op">=</span> <span class="st">"union"</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="st">"inboth"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"mclr"</span>, </span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.5</span>, </span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">3.8</span>,</span>
<span>     groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No seasonal allergies"</span>, <span class="st">"Seasonal allergies"</span><span class="op">)</span>,</span>
<span>     hubBorderCol  <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottom"</span>, title <span class="op">=</span> <span class="st">"estimated association:"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, </span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, inset <span class="op">=</span> <span class="fl">0.02</span>, cex <span class="op">=</span> <span class="fl">4</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">4</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/netcomp_spring_6-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="quantitative-network-comparison">Quantitative network comparison<a class="anchor" aria-label="anchor" href="#quantitative-network-comparison"></a></h4>
<p>Since runtime is considerably increased if permutation tests are performed, we set the <code>permTest</code> parameter to <code>FALSE</code>. See the <code>tutorial_createAssoPerm</code> file for a network comparison including permutation tests.</p>
<p>Since permutation tests are still conducted for the Adjusted Rand Index, a seed should be set for reproducibility.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comp_season</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netCompare.html">netCompare</a></span><span class="op">(</span><span class="va">props_season</span>, </span>
<span>                          permTest <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                          verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          seed <span class="op">=</span> <span class="fl">123456</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">comp_season</span>, </span>
<span>        groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No allergies"</span>, <span class="st">"Allergies"</span><span class="op">)</span>,</span>
<span>        showCentr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"degree"</span>, <span class="st">"between"</span>, <span class="st">"closeness"</span><span class="op">)</span>, </span>
<span>        numbNodes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a><span class="do">## Comparison of Network Properties</span></span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a><span class="do">## ----------------------------------</span></span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a><span class="do">## CALL: </span></span>
<span id="cb58-5"><a href="#cb58-5" tabindex="-1"></a><span class="do">## netCompare(x = props_season, permTest = FALSE, verbose = FALSE, </span></span>
<span id="cb58-6"><a href="#cb58-6" tabindex="-1"></a><span class="do">##     seed = 123456)</span></span>
<span id="cb58-7"><a href="#cb58-7" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-8"><a href="#cb58-8" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb58-9"><a href="#cb58-9" tabindex="-1"></a><span class="do">## Global network properties</span></span>
<span id="cb58-10"><a href="#cb58-10" tabindex="-1"></a><span class="do">## `````````````````````````</span></span>
<span id="cb58-11"><a href="#cb58-11" tabindex="-1"></a><span class="do">## Largest connected component (LCC):</span></span>
<span id="cb58-12"><a href="#cb58-12" tabindex="-1"></a><span class="do">##                          No allergies   Allergies    difference</span></span>
<span id="cb58-13"><a href="#cb58-13" tabindex="-1"></a><span class="do">## Relative LCC size               0.667       0.738         0.071</span></span>
<span id="cb58-14"><a href="#cb58-14" tabindex="-1"></a><span class="do">## Clustering coefficient          0.152       0.271         0.120</span></span>
<span id="cb58-15"><a href="#cb58-15" tabindex="-1"></a><span class="do">## Modularity                      0.626       0.458         0.168</span></span>
<span id="cb58-16"><a href="#cb58-16" tabindex="-1"></a><span class="do">## Positive edge percentage       86.667     100.000        13.333</span></span>
<span id="cb58-17"><a href="#cb58-17" tabindex="-1"></a><span class="do">## Edge density                    0.079       0.125         0.045</span></span>
<span id="cb58-18"><a href="#cb58-18" tabindex="-1"></a><span class="do">## Natural connectivity            0.045       0.044         0.002</span></span>
<span id="cb58-19"><a href="#cb58-19" tabindex="-1"></a><span class="do">## Vertex connectivity             1.000       1.000         0.000</span></span>
<span id="cb58-20"><a href="#cb58-20" tabindex="-1"></a><span class="do">## Edge connectivity               1.000       1.000         0.000</span></span>
<span id="cb58-21"><a href="#cb58-21" tabindex="-1"></a><span class="do">## Average dissimilarity*          0.673       0.682         0.009</span></span>
<span id="cb58-22"><a href="#cb58-22" tabindex="-1"></a><span class="do">## Average path length**           3.400       1.868         1.532</span></span>
<span id="cb58-23"><a href="#cb58-23" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-24"><a href="#cb58-24" tabindex="-1"></a><span class="do">## Whole network:</span></span>
<span id="cb58-25"><a href="#cb58-25" tabindex="-1"></a><span class="do">##                          No allergies   Allergies    difference</span></span>
<span id="cb58-26"><a href="#cb58-26" tabindex="-1"></a><span class="do">## Number of components           15.000       5.000        10.000</span></span>
<span id="cb58-27"><a href="#cb58-27" tabindex="-1"></a><span class="do">## Clustering coefficient          0.152       0.298         0.146</span></span>
<span id="cb58-28"><a href="#cb58-28" tabindex="-1"></a><span class="do">## Modularity                      0.626       0.557         0.069</span></span>
<span id="cb58-29"><a href="#cb58-29" tabindex="-1"></a><span class="do">## Positive edge percentage       86.667     100.000        13.333</span></span>
<span id="cb58-30"><a href="#cb58-30" tabindex="-1"></a><span class="do">## Edge density                    0.035       0.081         0.046</span></span>
<span id="cb58-31"><a href="#cb58-31" tabindex="-1"></a><span class="do">## Natural connectivity            0.028       0.031         0.003</span></span>
<span id="cb58-32"><a href="#cb58-32" tabindex="-1"></a><span class="do">## -----</span></span>
<span id="cb58-33"><a href="#cb58-33" tabindex="-1"></a><span class="do">##  *: Dissimilarity = 1 - edge weight</span></span>
<span id="cb58-34"><a href="#cb58-34" tabindex="-1"></a><span class="do">## **: Path length = Sum of dissimilarities along the path</span></span>
<span id="cb58-35"><a href="#cb58-35" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-36"><a href="#cb58-36" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb58-37"><a href="#cb58-37" tabindex="-1"></a><span class="do">## Jaccard index (similarity betw. sets of most central nodes)</span></span>
<span id="cb58-38"><a href="#cb58-38" tabindex="-1"></a><span class="do">## ```````````````````````````````````````````````````````````</span></span>
<span id="cb58-39"><a href="#cb58-39" tabindex="-1"></a><span class="do">##                     Jacc   P(&lt;=Jacc)     P(&gt;=Jacc)   </span></span>
<span id="cb58-40"><a href="#cb58-40" tabindex="-1"></a><span class="do">## degree             0.556    0.957578      0.144846   </span></span>
<span id="cb58-41"><a href="#cb58-41" tabindex="-1"></a><span class="do">## betweenness centr. 0.333    0.650307      0.622822   </span></span>
<span id="cb58-42"><a href="#cb58-42" tabindex="-1"></a><span class="do">## closeness centr.   0.231    0.322424      0.861268   </span></span>
<span id="cb58-43"><a href="#cb58-43" tabindex="-1"></a><span class="do">## eigenvec. centr.   0.100    0.017593 *    0.996692   </span></span>
<span id="cb58-44"><a href="#cb58-44" tabindex="-1"></a><span class="do">## hub taxa           0.000    0.296296      1.000000   </span></span>
<span id="cb58-45"><a href="#cb58-45" tabindex="-1"></a><span class="do">## -----</span></span>
<span id="cb58-46"><a href="#cb58-46" tabindex="-1"></a><span class="do">## Jaccard index in [0,1] (1 indicates perfect agreement)</span></span>
<span id="cb58-47"><a href="#cb58-47" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-48"><a href="#cb58-48" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb58-49"><a href="#cb58-49" tabindex="-1"></a><span class="do">## Adjusted Rand index (similarity betw. clusterings)</span></span>
<span id="cb58-50"><a href="#cb58-50" tabindex="-1"></a><span class="do">## ``````````````````````````````````````````````````</span></span>
<span id="cb58-51"><a href="#cb58-51" tabindex="-1"></a><span class="do">##         wholeNet       LCC</span></span>
<span id="cb58-52"><a href="#cb58-52" tabindex="-1"></a><span class="do">## ARI        0.232     0.355</span></span>
<span id="cb58-53"><a href="#cb58-53" tabindex="-1"></a><span class="do">## p-value    0.000     0.000</span></span>
<span id="cb58-54"><a href="#cb58-54" tabindex="-1"></a><span class="do">## -----</span></span>
<span id="cb58-55"><a href="#cb58-55" tabindex="-1"></a><span class="do">## ARI in [-1,1] with ARI=1: perfect agreement betw. clusterings</span></span>
<span id="cb58-56"><a href="#cb58-56" tabindex="-1"></a><span class="do">##                    ARI=0: expected for two random clusterings</span></span>
<span id="cb58-57"><a href="#cb58-57" tabindex="-1"></a><span class="do">## p-value: permutation test (n=1000) with null hypothesis ARI=0</span></span>
<span id="cb58-58"><a href="#cb58-58" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-59"><a href="#cb58-59" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb58-60"><a href="#cb58-60" tabindex="-1"></a><span class="do">## Graphlet Correlation Distance</span></span>
<span id="cb58-61"><a href="#cb58-61" tabindex="-1"></a><span class="do">## `````````````````````````````</span></span>
<span id="cb58-62"><a href="#cb58-62" tabindex="-1"></a><span class="do">##     wholeNet       LCC</span></span>
<span id="cb58-63"><a href="#cb58-63" tabindex="-1"></a><span class="do">## GCD    1.577     1.863</span></span>
<span id="cb58-64"><a href="#cb58-64" tabindex="-1"></a><span class="do">## -----</span></span>
<span id="cb58-65"><a href="#cb58-65" tabindex="-1"></a><span class="do">## GCD &gt;= 0 (GCD=0 indicates perfect agreement between GCMs)</span></span>
<span id="cb58-66"><a href="#cb58-66" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-67"><a href="#cb58-67" tabindex="-1"></a><span class="do">## ______________________________</span></span>
<span id="cb58-68"><a href="#cb58-68" tabindex="-1"></a><span class="do">## Centrality measures</span></span>
<span id="cb58-69"><a href="#cb58-69" tabindex="-1"></a><span class="do">## - In decreasing order</span></span>
<span id="cb58-70"><a href="#cb58-70" tabindex="-1"></a><span class="do">## - Computed for the whole network</span></span>
<span id="cb58-71"><a href="#cb58-71" tabindex="-1"></a><span class="do">## ````````````````````````````````````</span></span>
<span id="cb58-72"><a href="#cb58-72" tabindex="-1"></a><span class="do">## Degree (unnormalized):</span></span>
<span id="cb58-73"><a href="#cb58-73" tabindex="-1"></a><span class="do">##        No allergies Allergies abs.diff.</span></span>
<span id="cb58-74"><a href="#cb58-74" tabindex="-1"></a><span class="do">## 322235            3         9         6</span></span>
<span id="cb58-75"><a href="#cb58-75" tabindex="-1"></a><span class="do">## 363302            3         9         6</span></span>
<span id="cb58-76"><a href="#cb58-76" tabindex="-1"></a><span class="do">## 469709            0         4         4</span></span>
<span id="cb58-77"><a href="#cb58-77" tabindex="-1"></a><span class="do">## 158660            2         6         4</span></span>
<span id="cb58-78"><a href="#cb58-78" tabindex="-1"></a><span class="do">## 223059            0         4         4</span></span>
<span id="cb58-79"><a href="#cb58-79" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-80"><a href="#cb58-80" tabindex="-1"></a><span class="do">## Betweenness centrality (unnormalized):</span></span>
<span id="cb58-81"><a href="#cb58-81" tabindex="-1"></a><span class="do">##        No allergies Allergies abs.diff.</span></span>
<span id="cb58-82"><a href="#cb58-82" tabindex="-1"></a><span class="do">## 307981          231         0       231</span></span>
<span id="cb58-83"><a href="#cb58-83" tabindex="-1"></a><span class="do">## 331820          170         9       161</span></span>
<span id="cb58-84"><a href="#cb58-84" tabindex="-1"></a><span class="do">## 259569          137        34       103</span></span>
<span id="cb58-85"><a href="#cb58-85" tabindex="-1"></a><span class="do">## 158660          162        80        82</span></span>
<span id="cb58-86"><a href="#cb58-86" tabindex="-1"></a><span class="do">## 184983           92        12        80</span></span>
<span id="cb58-87"><a href="#cb58-87" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-88"><a href="#cb58-88" tabindex="-1"></a><span class="do">## Closeness centrality (unnormalized):</span></span>
<span id="cb58-89"><a href="#cb58-89" tabindex="-1"></a><span class="do">##        No allergies Allergies abs.diff.</span></span>
<span id="cb58-90"><a href="#cb58-90" tabindex="-1"></a><span class="do">## 469709            0    21.203    21.203</span></span>
<span id="cb58-91"><a href="#cb58-91" tabindex="-1"></a><span class="do">## 541301            0    20.942    20.942</span></span>
<span id="cb58-92"><a href="#cb58-92" tabindex="-1"></a><span class="do">## 181016            0    19.498    19.498</span></span>
<span id="cb58-93"><a href="#cb58-93" tabindex="-1"></a><span class="do">## 361496            0    19.349    19.349</span></span>
<span id="cb58-94"><a href="#cb58-94" tabindex="-1"></a><span class="do">## 223059            0    19.261    19.261</span></span>
<span id="cb58-95"><a href="#cb58-95" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb58-96"><a href="#cb58-96" tabindex="-1"></a><span class="do">## _________________________________________________________</span></span>
<span id="cb58-97"><a href="#cb58-97" tabindex="-1"></a><span class="do">## Significance codes: ***: 0.001, **: 0.01, *: 0.05, .: 0.1</span></span></code></pre></div>
<hr></div>
</div>
<div class="section level3">
<h3 id="differential-networks">Differential networks<a class="anchor" aria-label="anchor" href="#differential-networks"></a></h3>
<p>We now build a differential association network, where two nodes are connected if they are differentially associated between the two groups.</p>
<p>Due to its very short execution time, we use Pearson’s correlations for estimating associations between OTUs.</p>
<p>Fisher’s z-test is applied for identifying differentially correlated OTUs. Multiple testing adjustment is done by controlling the local false discovery rate.</p>
<p>Note: <code>sparsMethod</code> is set to <code>"none"</code>, just to be able to include all differential associations in the association network plot (see below). However, the differential network is always based on the estimated association matrices before sparsification (the <code>assoEst1</code> and <code>assoEst2</code> matrices returned by <code><a href="reference/netConstruct.html">netConstruct()</a></code>).</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_season_pears</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">amgut_season_no</span>, </span>
<span>                                 data2 <span class="op">=</span> <span class="va">amgut_season_yes</span>, </span>
<span>                                 filtTax <span class="op">=</span> <span class="st">"highestVar"</span>,</span>
<span>                                 filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestVar <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>                                 measure <span class="op">=</span> <span class="st">"pearson"</span>, </span>
<span>                                 normMethod <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                                 sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                                 thresh <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                                 verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="do">## Checking input arguments ... Done.</span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a><span class="do">## Infos about changed arguments:</span></span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a><span class="do">## Zero replacement needed for clr transformation. "multRepl" used.</span></span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a><span class="do">## Data filtering ...</span></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a><span class="do">## 95 taxa removed in each data set.</span></span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a><span class="do">## 1 rows with zero sum removed in group 1.</span></span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a><span class="do">## 1 rows with zero sum removed in group 2.</span></span>
<span id="cb60-9"><a href="#cb60-9" tabindex="-1"></a><span class="do">## 43 taxa and 162 samples remaining in group 1.</span></span>
<span id="cb60-10"><a href="#cb60-10" tabindex="-1"></a><span class="do">## 43 taxa and 120 samples remaining in group 2.</span></span>
<span id="cb60-11"><a href="#cb60-11" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-12"><a href="#cb60-12" tabindex="-1"></a><span class="do">## Zero treatment in group 1:</span></span>
<span id="cb60-13"><a href="#cb60-13" tabindex="-1"></a><span class="do">## Execute multRepl() ... Done.</span></span>
<span id="cb60-14"><a href="#cb60-14" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-15"><a href="#cb60-15" tabindex="-1"></a><span class="do">## Zero treatment in group 2:</span></span>
<span id="cb60-16"><a href="#cb60-16" tabindex="-1"></a><span class="do">## Execute multRepl() ... Done.</span></span>
<span id="cb60-17"><a href="#cb60-17" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-18"><a href="#cb60-18" tabindex="-1"></a><span class="do">## Normalization in group 1:</span></span>
<span id="cb60-19"><a href="#cb60-19" tabindex="-1"></a><span class="do">## Execute clr(){SpiecEasi} ... Done.</span></span>
<span id="cb60-20"><a href="#cb60-20" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-21"><a href="#cb60-21" tabindex="-1"></a><span class="do">## Normalization in group 2:</span></span>
<span id="cb60-22"><a href="#cb60-22" tabindex="-1"></a><span class="do">## Execute clr(){SpiecEasi} ... Done.</span></span>
<span id="cb60-23"><a href="#cb60-23" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-24"><a href="#cb60-24" tabindex="-1"></a><span class="do">## Calculate 'pearson' associations ... Done.</span></span>
<span id="cb60-25"><a href="#cb60-25" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb60-26"><a href="#cb60-26" tabindex="-1"></a><span class="do">## Calculate associations in group 2 ... Done.</span></span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differential network construction</span></span>
<span><span class="va">diff_season</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/diffnet.html">diffnet</a></span><span class="op">(</span><span class="va">net_season_pears</span>,</span>
<span>                       diffMethod <span class="op">=</span> <span class="st">"fisherTest"</span>, </span>
<span>                       adjust <span class="op">=</span> <span class="st">"lfdr"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb62"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="do">## Checking input arguments ... </span></span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a><span class="do">## Done.</span></span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a><span class="do">## Adjust for multiple testing using 'lfdr' ... </span></span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a><span class="do">## Execute fdrtool() ...</span></span>
<span id="cb62-5"><a href="#cb62-5" tabindex="-1"></a></span>
<span id="cb62-6"><a href="#cb62-6" tabindex="-1"></a><span class="do">## Step 1... determine cutoff point</span></span>
<span id="cb62-7"><a href="#cb62-7" tabindex="-1"></a><span class="do">## Step 2... estimate parameters of null distribution and eta0</span></span>
<span id="cb62-8"><a href="#cb62-8" tabindex="-1"></a><span class="do">## Step 3... compute p-values and estimate empirical PDF/CDF</span></span>
<span id="cb62-9"><a href="#cb62-9" tabindex="-1"></a><span class="do">## Step 4... compute q-values and local fdr</span></span>
<span id="cb62-10"><a href="#cb62-10" tabindex="-1"></a></span>
<span id="cb62-11"><a href="#cb62-11" tabindex="-1"></a><span class="do">## Done.</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differential network plot</span></span>
<span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">diff_season</span>, </span>
<span>     cexNodes <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>     cexLegend <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">4</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">8</span>,<span class="fl">5</span><span class="op">)</span>,</span>
<span>     legendGroupnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group 'no'"</span>, <span class="st">"group 'yes'"</span><span class="op">)</span>,</span>
<span>     legendPos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.7</span>,<span class="fl">1.6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/diffnet_1-1.png"><!-- --> In the differential network shown above, edge colors represent the direction of associations in the two groups. If, for instance, two OTUs are positively associated in group 1 and negatively associated in group 2 (such as ‘191541’ and ‘188236’), the respective edge is colored in cyan.</p>
<p>We also take a look at the corresponding associations by constructing association networks that include only the differentially associated OTUs.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_season_pears</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_season_pears</span>, </span>
<span>                                 clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                                 weightDeg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                 normDeg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                 gcmHeat <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify the differentially associated OTUs</span></span>
<span><span class="va">diffmat_sums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">diff_season</span><span class="op">$</span><span class="va">diffAdjustMat</span><span class="op">)</span></span>
<span><span class="va">diff_asso_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">diffmat_sums</span><span class="op">[</span><span class="va">diffmat_sums</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_season_pears</span>, </span>
<span>     nodeFilter <span class="op">=</span> <span class="st">"names"</span>,</span>
<span>     nodeFilterPar <span class="op">=</span> <span class="va">diff_asso_names</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"gray"</span>,</span>
<span>     highlightHubs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     sameLayout <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>     layoutGroup <span class="op">=</span> <span class="st">"union"</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.5</span>, </span>
<span>     cexLabels <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">3.8</span>,</span>
<span>     groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No seasonal allergies"</span>, <span class="st">"Seasonal allergies"</span><span class="op">)</span>,</span>
<span>     hubBorderCol  <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>,<span class="op">-</span><span class="fl">0.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, </span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, inset <span class="op">=</span> <span class="fl">0.05</span>, cex <span class="op">=</span> <span class="fl">4</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">4</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/diffnet_3-1.png"><!-- --></p>
<p>We can see that the correlation between the aforementioned OTUs ‘191541’ and ‘188236’ is strongly positive in the left group and negative in the right group.</p>
<hr></div>
<div class="section level3">
<h3 id="dissimilarity-based-networks">Dissimilarity-based Networks<a class="anchor" aria-label="anchor" href="#dissimilarity-based-networks"></a></h3>
<p>If a dissimilarity measure is used for network construction, nodes are subjects instead of OTUs. The estimated dissimilarities are transformed into similarities, which are used as edge weights so that subjects with a similar microbial composition are placed close together in the network plot.</p>
<p>We construct a single network using Aitchison’s distance being suitable for the application on compositional data.</p>
<p>Since the Aitchison distance is based on the clr-transformation, zeros in the data need to be replaced.</p>
<p>The network is sparsified using the k-nearest neighbor (knn) algorithm.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_diss</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut1.filt</span>,</span>
<span>                         measure <span class="op">=</span> <span class="st">"aitchison"</span>,</span>
<span>                         zeroMethod <span class="op">=</span> <span class="st">"multRepl"</span>,</span>
<span>                         sparsMethod <span class="op">=</span> <span class="st">"knn"</span>,</span>
<span>                         kNeighbor <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb67"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="do">## Checking input arguments ... Done.</span></span>
<span id="cb67-2"><a href="#cb67-2" tabindex="-1"></a><span class="do">## Infos about changed arguments:</span></span>
<span id="cb67-3"><a href="#cb67-3" tabindex="-1"></a><span class="do">## Counts normalized to fractions for measure "aitchison".</span></span>
<span id="cb67-4"><a href="#cb67-4" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb67-5"><a href="#cb67-5" tabindex="-1"></a><span class="do">## 127 taxa and 289 samples remaining.</span></span>
<span id="cb67-6"><a href="#cb67-6" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb67-7"><a href="#cb67-7" tabindex="-1"></a><span class="do">## Zero treatment:</span></span>
<span id="cb67-8"><a href="#cb67-8" tabindex="-1"></a><span class="do">## Execute multRepl() ... Done.</span></span>
<span id="cb67-9"><a href="#cb67-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb67-10"><a href="#cb67-10" tabindex="-1"></a><span class="do">## Normalization:</span></span>
<span id="cb67-11"><a href="#cb67-11" tabindex="-1"></a><span class="do">## Counts normalized by total sum scaling.</span></span>
<span id="cb67-12"><a href="#cb67-12" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb67-13"><a href="#cb67-13" tabindex="-1"></a><span class="do">## Calculate 'aitchison' dissimilarities ... Done.</span></span>
<span id="cb67-14"><a href="#cb67-14" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb67-15"><a href="#cb67-15" tabindex="-1"></a><span class="do">## Sparsify dissimilarities via 'knn' ... Registered S3 methods overwritten by 'proxy':</span></span>
<span id="cb67-16"><a href="#cb67-16" tabindex="-1"></a><span class="do">##   method               from    </span></span>
<span id="cb67-17"><a href="#cb67-17" tabindex="-1"></a><span class="do">##   print.registry_field registry</span></span>
<span id="cb67-18"><a href="#cb67-18" tabindex="-1"></a><span class="do">##   print.registry_entry registry</span></span>
<span id="cb67-19"><a href="#cb67-19" tabindex="-1"></a><span class="do">## Done.</span></span></code></pre></div>
<p>For cluster detection, we use hierarchical clustering with average linkage. Internally, <code>k=3</code> is passed to <a href="https://www.rdocumentation.org/packages/dendextend/versions/1.13.4/topics/cutree" class="external-link"><code>cutree()</code></a> from <code>stats</code> package so that the tree is cut into 3 clusters.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_diss</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_diss</span>,</span>
<span>                         clustMethod <span class="op">=</span> <span class="st">"hierarchical"</span>,</span>
<span>                         clustPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"average"</span>, k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,</span>
<span>                         hubPar <span class="op">=</span> <span class="st">"eigenvector"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/example14-1.png"><!-- --></p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_diss</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     hubTransp <span class="op">=</span> <span class="fl">40</span>,</span>
<span>     edgeTranspLow <span class="op">=</span> <span class="fl">60</span>,</span>
<span>     charToRm <span class="op">=</span> <span class="st">"00000"</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"simple"</span>,</span>
<span>     labelLength <span class="op">=</span> <span class="fl">6</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get green color with 50% transparency</span></span>
<span><span class="va">green2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/colToTransp.html">colToTransp</a></span><span class="op">(</span><span class="st">"#009900"</span>, <span class="fl">40</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">1.1</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">2.2</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"high similarity (low Aitchison distance)"</span>,</span>
<span>                  <span class="st">"low similarity (high Aitchison distance)"</span><span class="op">)</span>, </span>
<span>       lty <span class="op">=</span> <span class="fl">1</span>, </span>
<span>       lwd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkgreen"</span>, <span class="va">green2</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/readme/example15-1.png"><!-- --></p>
<p>In this dissimilarity-based network, hubs are interpreted as samples with a microbial composition similar to that of many other samples in the data set.</p>
<hr></div>
<div class="section level3">
<h3 id="soil-microbiome-example">Soil microbiome example<a class="anchor" aria-label="anchor" href="#soil-microbiome-example"></a></h3>
<p>Here is the code for reproducing the network plot shown at the beginning.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"soilrep"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">soil_warm_yes</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">soilrep</span>, <span class="va">warmed</span> <span class="op">==</span> <span class="st">"yes"</span><span class="op">)</span></span>
<span><span class="va">soil_warm_no</span>  <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">soilrep</span>, <span class="va">warmed</span> <span class="op">==</span> <span class="st">"no"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">net_seas_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">soil_warm_yes</span>, <span class="va">soil_warm_no</span>,</span>
<span>                           filtTax <span class="op">=</span> <span class="st">"highestVar"</span>,</span>
<span>                           filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestVar <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>,</span>
<span>                           zeroMethod <span class="op">=</span> <span class="st">"pseudo"</span>,</span>
<span>                           normMethod <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                           measure <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">netprops1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_seas_p</span>, clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nclust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">netprops1</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clust1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">topo.colors</a></span><span class="op">(</span><span class="va">nclust</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">netprops1</span>, </span>
<span>     sameLayout <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>     layoutGroup <span class="op">=</span> <span class="st">"union"</span>, </span>
<span>     colorVec <span class="op">=</span> <span class="va">col</span>,</span>
<span>     borderCol <span class="op">=</span> <span class="st">"gray40"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"degree"</span>, </span>
<span>     cexNodes <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>, </span>
<span>     edgeTranspLow <span class="op">=</span> <span class="fl">80</span>, </span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">50</span>,</span>
<span>     groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Warming"</span>, <span class="st">"Non-warming"</span><span class="op">)</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.8</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span>, </span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>     labels <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>     rmSingles <span class="op">=</span> <span class="st">"inboth"</span>,</span>
<span>     nodeFilter <span class="op">=</span> <span class="st">"clustMin"</span>, </span>
<span>     nodeFilterPar <span class="op">=</span> <span class="fl">10</span>, </span>
<span>     nodeTransp <span class="op">=</span> <span class="fl">50</span>, </span>
<span>     hubTransp <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<hr></div>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-badri2020shrinkage" class="csl-entry">
Badri, Michelle, Zachary D. Kurtz, Richard Bonneau, and Christian L. Müller. 2020. “Shrinkage Improves Estimation of Microbial Associations Under Different Normalization Methods.” <em>NAR Genomics and Bioinformatics</em> 2 (December). <a href="https://doi.org/10.1093/NARGAB/LQAA100" class="external-link uri">https://doi.org/10.1093/NARGAB/LQAA100</a>.
</div>
<div id="ref-martin1999measure" class="csl-entry">
Martin-Fernández, Josep A, M Bren, Carles Barceló-Vidal, and Vera Pawlowsky-Glahn. 1999. “A Measure of Difference for Compositional Data Based on Measures of Divergence.” In <em>Proceedings of IAMG</em>, 99:211–16.
</div>
<div id="ref-yoon2020fast" class="csl-entry">
Yoon, Grace, Christian L. Müller, and Irina Gaynanova. 2020. “Fast Computation of Latent Correlations.” <em>Journal of Computational and Graphical Statistics</em>, June. <a href="http://arxiv.org/abs/2006.13875" class="external-link uri">http://arxiv.org/abs/2006.13875</a>.
</div>
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://stefpeschel.de/" class="external-link">Stefanie Peschel</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

