<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • NetCoMi</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://netcomi.de/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">NetCoMi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/NetCoMi.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/net_comparison.html">Network comparison</a></li>
    <li><a class="dropdown-item" href="../articles/diffnet.html">Differential networks</a></li>
    <li><a class="dropdown-item" href="../articles/dissimilarity_networks.html">Dissimilarity-based Networks</a></li>
    <li><a class="dropdown-item" href="../articles/create_asso_perm.html">Generate permuted association matrices</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>More examples</h6></li>
    <li><a class="dropdown-item" href="../articles/plot_modifications.html">Modifying the network plot</a></li>
    <li><a class="dropdown-item" href="../articles/asso_as_input.html">Association matrix as input</a></li>
    <li><a class="dropdown-item" href="../articles/cross_domain_networks.html">Cross-domain networks</a></li>
    <li><a class="dropdown-item" href="../articles/soil_example.html">Soil microbiome example</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/stefpeschel/NetCoMi/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/stefpeschel/NetCoMi/blob/main/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.0" id="netcomi-120">NetCoMi 1.2.0<a class="anchor" aria-label="anchor" href="#netcomi-120"></a></h2>
<div class="section level3">
<h3 id="new-features-1-2-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-2-0"></a></h3>
<ul><li><p><code><a href="../reference/netConstruct.html">netConstruct()</a></code> now returns a <strong>message</strong> if the constructed network has <strong>no edges</strong>.</p></li>
<li><p>In <code><a href="../reference/netConstruct.html">netConstruct()</a></code>: The <strong>sparse + low-rank decomposition (slr)</strong> implemented in SpiecEasi (latentcor branch) can now be used with NetCoMi. See section “<a href="https://github.com/zdk123/SpiecEasi/tree/latentcor?tab=readme-ov-file#learning-latent-variable-graphical-models" class="external-link">Learning latent variable graphical models</a>” on SpiecEasi’s github page for details.</p></li>
<li><p>In <code><a href="../reference/netConstruct.html">netConstruct()</a></code>: The output of the function used for association/dissimilarity estimation (defined by <code>measure</code>) is now returned by <code>netConstruct</code> (objects <code>measureOut1</code> and <code>measureOut2</code>).</p></li>
<li><p><code><a href="../reference/netConstruct.html">netConstruct()</a></code> now accepts data of type <strong>“TreeSummarizedExperiment”</strong> as input to be compatible with the miaverse. Details can be found here: <a href="https://microbiome.github.io/OMA/docs/devel/pages/miaverse.html" class="external-link uri">https://microbiome.github.io/OMA/docs/devel/pages/miaverse.html</a></p></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-2-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-2-0"></a></h3>
<ul><li><p>In <code>netAnalyze</code>: Since the <code>scale</code> argument of the igraph function <code>eigen_centrality</code> is deprecated, NetCoMi’s <code>normEigen</code> argument is deprecated as well. Eigenvector centrality is now always scaled/normalized.</p></li>
<li><p>In <code>netAnalyze</code>: Since the igraph function <code>get.edgelist</code> is deprecated, <code>as_edgelist</code> is used instead.</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.0" id="netcomi-110">NetCoMi 1.1.0<a class="anchor" aria-label="anchor" href="#netcomi-110"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-1-0"></a></h3>
<ul><li><p><strong><code><a href="../reference/renameTaxa.html">renameTaxa()</a></code></strong>: New function for renaming taxa in a taxonomic table. It comes with functionality for making unknown and unclassified taxa unique and substituting them by the next higher known taxonomic level. E.g., an unknown genus “g__“, where family is the next higher known level, can automatically be renamed to”1_Streptococcaceae(F)“. User-defined patterns determine the format of known and substituted names. Unknown names (e.g., NAs) and unclassified taxa can be handled separately. Duplicated names within one or more chosen ranks can also be made unique by numbering them consecutively.</p></li>
<li><p><strong><code><a href="../reference/editLabels.html">editLabels()</a></code></strong>: New function for editing node labels, i.e., shortening to a certain length and removing unwanted characters. It is used by NetCoMi’s plot functions plot.microNetProps() and plot.diffnet().</p></li>
<li><p>In <code><a href="../reference/netCompare.html">netCompare()</a></code>: The <strong>adjusted Rand index</strong> is also computed for the largest connected component (LCC). The summary method has been adapted.</p></li>
<li><p>Argument <strong>“testRand”</strong> added to <code><a href="../reference/netCompare.html">netCompare()</a></code>. Performing a permutation test for the adjusted Rand index can now be disabled to save run time.</p></li>
<li><p><strong>Graphlet-based network measures</strong> implemented. NetCoMi contains two new exported functions <strong><code><a href="../reference/calcGCM.html">calcGCM()</a></code></strong> and <strong><code><a href="../reference/calcGCD.html">calcGCD()</a></code></strong> to compute the Graphlet Correlation Matrix (GCM) of a network and the Graphlet Correlation Distance (GCD) between two networks. <strong>Orbits</strong> for graphlets with up to four nodes are considered. Furthermore, the GCM is computed with <code><a href="../reference/netAnalyze.html">netAnalyze()</a></code> and the GCD with <code><a href="../reference/netCompare.html">netCompare()</a></code> (for the whole network and the largest connected component, respectively). Also the orbit counts are returned. The GCD is added to the summary for class <code>microNetComp</code> objects returned by <code><a href="../reference/netCompare.html">netCompare()</a></code>.</p></li>
<li><p><strong>Significance test for the GCD</strong>: If permutation tests are conducted with <code><a href="../reference/netCompare.html">netCompare()</a></code>, the GCD is tested for being significantly different from zero.</p></li>
<li><p>New function <strong><code><a href="../reference/testGCM.html">testGCM()</a></code></strong> to <strong>test graphlet-based measures</strong> for significance. For a single GCM, the correlations are tested for being significantly different from zero. If two GCMs are given, it is tested if the correlations are significantly different between the two groups, that is, the absolute differences between correlations ( <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">|</mo><mi>g</mi><mi>c</mi><msub><mn>1</mn><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>−</mo><mi>g</mi><mi>c</mi><msub><mn>2</mn><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">|</mo></mrow><annotation encoding="application/x-tex">|gc1_{ij}-gc2_{ij}|</annotation></semantics></math> ) are tested for being different from zero.</p></li>
<li><p>New function <strong><code><a href="../reference/plotHeat.html">plotHeat()</a></code></strong> for plotting a mixed heatmap where, for instance, values are shown in the upper triangle and corresponding p-values or significance codes in the lower triangle. The function is used for plotting heatmaps of the GCMs, but could also be used for association matrices.</p></li>
<li><p><code><a href="../reference/netAnalyze.html">netAnalyze()</a></code> now by default returns a <strong>heatmap of the GCM(s)</strong> with graphlet correlations in the upper triangle and significance codes in the lower triangle.</p></li>
<li><p>Argument <strong>“doPlot”</strong> added to <code><a href="../reference/plot.microNetProps.html">plot.microNetProps()</a></code> to suppress the plot if only the return value is of interest.</p></li>
<li><p>New <strong>“show”</strong> arguments are added to the summary methods for class <code>microNetProps</code> and <code>microNetComp</code> objects. They specify which network properties should be printed in the summary. See the help pages of <code>summary.microNetProps</code> and <code><a href="../reference/summarize.microNetComp.html">summary.microNetComp()</a></code> for details.</p></li>
<li><p>New <strong>zero replacement</strong> method <strong>“pseudoZO”</strong> available in <code><a href="../reference/netConstruct.html">netConstruct()</a></code>. Instead of adding the desired pseudo count to the whole count matrix, it is added to zero counts only if <code>pseudoZO</code> is chosen. The behavior of “pseudo” (a further available method where a pseudo count is added to all counts) has not changed. Adding a pseudo count only to zeros preserves the ratios between non-zero counts, which is desirable.</p></li>
<li><p><code><a href="../reference/createAssoPerm.html">createAssoPerm()</a></code> now accepts objects of class <code>microNet</code> as input (in addition to objects of class <code>microNetProps</code>).</p></li>
<li><p><strong><code>SPRING's</code></strong> fast version of latent correlation computation (implemented in <a href="https://github.com/irinagain/mixedCCA" class="external-link">mixedCCA</a>) is available again. It can be used by setting the <code><a href="../reference/netConstruct.html">netConstruct()</a></code> parameter <code>measurePar$Rmethod</code> to “approx”, which is now the default again.</p></li>
<li><p>The function <strong><code><a href="../reference/multAdjust.html">multAdjust()</a></code></strong> now has an argument <code>pTrueNull</code> to pre-define the proportion of true null hypotheses for the adaptive BH method.</p></li>
<li><p><code><a href="../reference/netConstruct.html">netConstruct()</a></code> has a new argument <strong><code>assoBoot</code></strong>, which enables the computation of bootstrap association matrices outside netConstruct() if <strong>bootstrapping</strong> is used for sparsification. An example has been added to the help page <code><a href="../reference/netConstruct.html">?netConstruct</a></code>. This feature might be useful for very large association matrices (for which the working memory might reach its limit).</p></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-1-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-0"></a></h3>
<ul><li>
<p>In <code><a href="../reference/netConstruct.html">netConstruct()</a></code>:</p>
<ul><li>Using <strong>“bootstrap”</strong> as sparsification method in combination with one of the association methods “bicor”, “cclasso”, “ccrepe”, or “gcoda” led to the error: <code>argument "verbose" is missing, with no default</code>, which has been fixed.</li>
<li>The <strong>“signedPos”</strong> transformation did not work properly. Dissimilarities corresponding to negative correlations were set to zero instead of infinity.</li>
</ul></li>
<li><p>In <code><a href="../reference/editLabels.html">editLabels()</a></code>: The function (and thus also <code>plot.microNetProps</code>) threw an error if taxa have been renamed with <code>renameTaxa</code> and the data contain more than 9 taxa with equal names, so that double-digit numbers were added to avoid duplicates.</p></li>
<li><p>Issues in network analysis and plotting if association matrices are used for network construction, but <strong>row and/or column names are missing</strong>. (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/65" class="external-link">#65</a>)</p></li>
<li><p><code><a href="../reference/diffnet.html">diffnet()</a></code> threw an error if association matrices are used for network construction instead of count matrices. (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/66" class="external-link">#66</a>)</p></li>
<li>
<p>In <code><a href="../reference/plot.microNetProps.html">plot.microNetProps()</a></code>:</p>
<ul><li>The function now directly returns an error if <code>x</code> has not the expected class.</li>
<li>The <code>cut</code> parameter could not be changed.</li>
</ul></li>
<li><p>In <strong><code><a href="../reference/cclasso.html">cclasso()</a></code></strong>: In rare cases, the function produced complex numbers, which led to an error.</p></li>
</ul></div>
<div class="section level3">
<h3 id="further-changes-1-1-0">Further changes<a class="anchor" aria-label="anchor" href="#further-changes-1-1-0"></a></h3>
<ul><li><p>In <strong>permutation tests</strong>: The permuted group labels must now be different from the original group vector. In other words, the original group vector is strictly avoided in the matrix with permuted group labels. So far, only duplicates were avoided. Only in exact permutation tests (if <code>nPerm</code> equals the possible number of permutations), the original group vector is still included in the permutation matrix. The calculation of p-values has been adapted to the new behavior: <em>p=B/N</em> for exact p-values and <em>p=(B+1)/(N+1)</em> for approximated p-values, where <em>B</em> is the number of permutation test statistics being larger than or equal to the observed one, and <em>N</em> is the number of permutations. So far, <em>p=(B+1)/(N+1)</em> has been used in all cases.</p></li>
<li>
<p>In <code><a href="../reference/plot.microNetProps.html">plot.microNetProps()</a></code>:</p>
<ul><li>The default of <code>shortenLabels</code> is now “none”, i.e. the <strong>labels are not shortened by default</strong>, to avoid confusion about the node labels.</li>
<li>The <strong>edge filter</strong> (specified via <code>edgeFilter</code> and <code>edgeInvisFilter</code>) now refers to the estimated association/dissimilarities instead of edge weights. E.g., setting the threshold to 0.3 for an association network hides edges with a corresponding absolute association below 0.3 even though the edge weight might be different (depending on the transformation used for network construction). (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/26" class="external-link">#26</a>)</li>
<li>If two networks are constructed and the <strong><code>cut</code></strong> parameter is not user-defined, the mean of the two determined cut parameters is now used for both networks so that edge thicknesses are comparable.</li>
</ul></li>
<li><p>More expressive messages and errors in <code>diffnet</code> and <code>plot.diffnet</code> if no <strong>differential associations</strong> are detected.</p></li>
<li><p>New function <strong><code>.suppress_warnings()</code></strong> to suppress certain warnings returned by external functions.</p></li>
<li><p>In <code>netConstruct</code> if <strong>“multRepl”</strong> is used for zero handling: The warning about the proportion of zeros is suppressed by setting the <code>multRepl()</code> parameter “z.warning” to 1.</p></li>
<li><p>The functions <strong><code>makeCluster</code></strong> and <strong><code>stopCluster</code></strong> from <code>parallel</code> package are now used for parallel computation because those from <code>snow</code> package sometimes led to problems on Unix machines.</p></li>
</ul></div>
<div class="section level3">
<h3 id="style-1-1-0">Style<a class="anchor" aria-label="anchor" href="#style-1-1-0"></a></h3>
<ul><li><p>The whole R code has been reformatted to follow general conventions.</p></li>
<li><p>The element <code>"clustering_lcc"</code> as part of the <code>netAnalyze</code> output has changed to <code>"clusteringLCC"</code> to be in line with the remaining output.</p></li>
<li><p>Input argument checking of exported function has been revised. New functions <code>.checkArgsXxx()</code> are added to perform argument checking outside the main functions.</p></li>
<li>
<p>Non-exported functions have been renamed to follow general naming conventions, i.e. that of <a href="https://contributions.bioconductor.org/r-code.html" class="external-link">Bioconductor</a>:</p>
<ul><li>Use camelCase for all functions.</li>
<li>Non-exported functions have prefix “.”</li>
<li>The following functions have been renamed:</li>
</ul></li>
</ul><table class="table"><thead><tr class="header"><th align="left">Old names</th>
<th align="left">New names</th>
</tr></thead><tbody><tr class="odd"><td align="left">boottest</td>
<td align="left">.boottest</td>
</tr><tr class="even"><td align="left">calc_association</td>
<td align="left">.calcAssociation</td>
</tr><tr class="odd"><td align="left">calc_diff_props</td>
<td align="left">.calcDiffProps</td>
</tr><tr class="even"><td align="left">calc_jaccard</td>
<td align="left">.calcJaccard</td>
</tr><tr class="odd"><td align="left">calc_props</td>
<td align="left">.calcProps</td>
</tr><tr class="even"><td align="left">diff_connect_pairs</td>
<td align="left">.diffConnectPairs</td>
</tr><tr class="odd"><td align="left">diff_connect_variables</td>
<td align="left">.diffConnectVariables</td>
</tr><tr class="even"><td align="left">diff_connect_network</td>
<td align="left">.diffConnectNetwork</td>
</tr><tr class="odd"><td align="left">filter_edges</td>
<td align="left">.filterEdges</td>
</tr><tr class="even"><td align="left">filter_nodes</td>
<td align="left">.filterNodes</td>
</tr><tr class="odd"><td align="left">filter_samples</td>
<td align="left">.filterSamples</td>
</tr><tr class="even"><td align="left">filter_taxa</td>
<td align="left">.filterTaxa</td>
</tr><tr class="odd"><td align="left">first_unequal_element</td>
<td align="left">.firstUnequalElement</td>
</tr><tr class="even"><td align="left">get_clust_cols</td>
<td align="left">.getClustCols</td>
</tr><tr class="odd"><td align="left">get_node_size</td>
<td align="left">.getNodeSize</td>
</tr><tr class="even"><td align="left">get_perm_group_mat</td>
<td align="left">.getPermGroupMat</td>
</tr><tr class="odd"><td align="left">get_vec_names</td>
<td align="left">.getVecNames</td>
</tr><tr class="even"><td align="left">norm_counts</td>
<td align="left">.normCounts</td>
</tr><tr class="odd"><td align="left">permtest_diff_asso</td>
<td align="left">.permTestDiffAsso</td>
</tr><tr class="even"><td align="left">scale_diss</td>
<td align="left">.scaleDiss</td>
</tr><tr class="odd"><td align="left">sparsify</td>
<td align="left">.sparsify</td>
</tr><tr class="even"><td align="left">trans_to_diss</td>
<td align="left">.transToDiss</td>
</tr><tr class="odd"><td align="left">trans_to_sim</td>
<td align="left">.transToSim</td>
</tr><tr class="even"><td align="left">trans_to_adja</td>
<td align="left">.transToAdja</td>
</tr><tr class="odd"><td align="left">zero_treat</td>
<td align="left">.zeroTreat</td>
</tr></tbody></table></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.3" id="netcomi-103">NetCoMi 1.0.3<a class="anchor" aria-label="anchor" href="#netcomi-103"></a></h2>
<p>This is a minor release with some bug fixes and changes in the documentation.</p>
<div class="section level3">
<h3 id="bug-fixes-1-0-3">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-0-3"></a></h3>
<ul><li><p><code><a href="../reference/netConstruct.html">netConstruct()</a></code> threw an error if the data had no row and/or column names, which is fixed.</p></li>
<li><p>An edge list is added to the output of <code><a href="../reference/netConstruct.html">netConstruct()</a></code> (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/41" class="external-link">#41</a>). See the help page for details.</p></li>
<li><p><code>SPRING</code>’s fast version of latent correlation computation (implemented in <a href="https://github.com/irinagain/mixedCCA" class="external-link">mixedCCA</a>) is currently not available due to deprecation of the R package <code>chebpol</code>. The issue is fixed by setting the <code><a href="../reference/netConstruct.html">netConstruct()</a></code> parameter <code>measurePar$Rmethod</code> internally to “original” if SPRING is used for association estimation.</p></li>
<li><p>In <code><a href="../reference/plot.microNetProps.html">plot.microNetProps()</a></code>: The <code>xpd</code> parameter is changed to <code>NA</code> so that plotting outside the plot region is possible (useful for legends or additional text).</p></li>
<li><p>Labels in the network plot can now be suppressed by setting <code>labels = FALSE</code> (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/43" class="external-link">#43</a>)</p></li>
<li><p>The <code><a href="../reference/netCompare.html">netCompare()</a></code> function threw an error if one of the permutation networks was empty, i.e. had no edges with weight different from zero (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/38" class="external-link">#38</a>), which is now fixed.</p></li>
<li><p>Fix issues <a href="https://github.com/stefpeschel/NetCoMi/issues/29" class="external-link">#29</a> and <a href="https://github.com/stefpeschel/NetCoMi/issues/40" class="external-link">#40</a>, where permutation tests did not terminate for small sample sizes. Now, if the possible number of permutations (resulting from the sample size) is smaller than that defined by the user, the function stops and returns an error.</p></li>
<li><p>Fix a bug in <code><a href="../reference/diffnet.html">diffnet()</a></code> (issue <a href="https://github.com/stefpeschel/NetCoMi/issues/51" class="external-link">#51</a>), where colors in differential networks could not be changed.</p></li>
<li><p><code><a href="../reference/diffnet.html">diffnet()</a></code> threw an error if the <code><a href="../reference/netConstruct.html">netConstruct()</a></code> argument <code>jointPrepro</code> was set to <code>TRUE</code>.</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.2" id="netcomi-102">NetCoMi 1.0.2<a class="anchor" aria-label="anchor" href="#netcomi-102"></a></h2>
<p>This release includes a range of new features and fixes known bugs and issues.</p>
<div class="section level3">
<h3 id="new-features-1-0-2">New features<a class="anchor" aria-label="anchor" href="#new-features-1-0-2"></a></h3>
<div class="section level4">
<h4 id="improved-installation-process-1-0-2">Improved installation process<a class="anchor" aria-label="anchor" href="#improved-installation-process-1-0-2"></a></h4>
<p>Packages that are optionally required in certain settings are not installed together with <code>NetCoMi</code> anymore. Instead, there is a new function <code><a href="../reference/installNetCoMiPacks.html">installNetCoMiPacks()</a></code> for installing the remaining packages. If not installed via <code><a href="../reference/installNetCoMiPacks.html">installNetCoMiPacks()</a></code>, the required package is installed by the respective NetCoMi function when needed.</p>
</div>
<div class="section level4">
<h4 id="installnetcomipacks-1-0-2">installNetCoMiPacks()<a class="anchor" aria-label="anchor" href="#installnetcomipacks-1-0-2"></a></h4>
<p>New function for installing the R packages used in NetCoMi not listed as <code>dependencies</code> or <code>imports</code> in NetCoMi’s description file.</p>
</div>
<div class="section level4">
<h4 id="netconstruct-1-0-2">netConstruct()<a class="anchor" aria-label="anchor" href="#netconstruct-1-0-2"></a></h4>
<ul><li><p>New argument <code>matchDesign</code>: Implements matched-group (i.e. matched-pair) designs, which are used for permutation tests in <code><a href="../reference/netCompare.html">netCompare()</a></code> and <code><a href="../reference/diffnet.html">diffnet()</a></code>. <code>c(1,2)</code>, for instance, means that one sample in the first group is matched to two samples in the second group. If the argument is not <code>NULL</code>, the matched-group design is kept when generating permuted data.</p></li>
<li><p>New argument <code>jointPrepro</code>: Specifies whether two data sets (of group one and two) should be preprocessed together. Preprocessing includes sample and taxa filtering, zero treatment, and normalization. Defaults to <code>TRUE</code> if <code>data</code> and <code>group</code> are given, and to <code>FALSE</code> if <code>data</code> and <code>data2</code> are given, which is similar to the behavior of <code>NetCoMi 1.0.1</code>. For dissimilarity networks, no joint preprocessing is possible.</p></li>
<li><p><code>mclr(){SPRING}</code> is now available as normalization method.</p></li>
<li><p><code>clr{SpiecEasi}</code> is used for centered log-ratio transformation instead of <code>cenLR(){robCompositions}</code>.</p></li>
<li><p><code>"symBetaMode"</code> is accepted as list element of <code>measurePar</code>, which is passed to <code>symBeta(){SpiecEasi}</code>. Only needed for SpiecEasi or SPRING associations.</p></li>
<li><p>The pseudocount (if <code>zeroMethod = "pseudo"</code>) may be freely specified. In v1.0.1, only unit pseudocounts were possible.</p></li>
</ul></div>
<div class="section level4">
<h4 id="netanalyze-1-0-2">netAnalyze()<a class="anchor" aria-label="anchor" href="#netanalyze-1-0-2"></a></h4>
<ul><li><p>Global network properties are now computed for the whole network as well as for the largest connected component (LCC). The summary of network properties now contains for the whole network only statistics that are not based on shortest paths (or, more generally, also meaningful for disconnected networks). For the LCC, all global properties available in NetCoMi are shown.</p></li>
<li>
<p>New global network properties (see the docu of <code><a href="../reference/netAnalyze.html">netAnalyze()</a></code> for definitions):</p>
<ul><li>Number of components (only whole network)</li>
<li>Relative LCC size (only LCC)</li>
<li>Positive edge percentage</li>
<li>Natural connectivity</li>
<li>Average dissimilarity (only meaningful for the LCC)</li>
<li>Average path length (only meaningful for the LCC)</li>
</ul></li>
<li><p>New argument <code>centrLCC</code>: Specifies whether to compute centralities only for the LCC. If <code>TRUE</code>, centrality values of disconnected components are zero.</p></li>
<li><p>New argument <code>avDissIgnoreInf</code>: Indicates whether infinite values should be ignored in the average dissimilarity. If <code>FALSE</code>, infinities are set to 1.</p></li>
<li><p>New argument <code>sPathAlgo</code>: Algorithm used for computing shortest paths</p></li>
<li><p>New argument <code>sPathNorm</code>: Indicates whether shortest paths should be normalized by average dissimilarity to improve interpretability.</p></li>
<li><p>New argument <code>normNatConnect</code>: Indicates whether to normalize natural connectivity values.</p></li>
<li><p>New argument <code>weightClustCoef</code>: Specifies the algorithm used for computing the global clustering coefficient. If <code>FALSE</code>, <code>transitivity(){igraph}</code> with <code>type = "global"</code> is used (similar to <code>NetCoMi 1.0.1</code>). If <code>TRUE</code>, the local clustering coefficient is computed using <code>transitivity(){igraph}</code> with <code>type = "barrat"</code>. The global clustering coefficient is then the arithmetic mean of local values.</p></li>
<li><p>Argument <code>connect</code> has been changed to <code>connectivity</code>.</p></li>
<li><p>Documentation extended by definitions of network properties.</p></li>
</ul></div>
<div class="section level4">
<h4 id="summarymicronetprops-1-0-2">summary.microNetProps()<a class="anchor" aria-label="anchor" href="#summarymicronetprops-1-0-2"></a></h4>
<ul><li><p>New argument <code>clusterLCC</code>: Indicates whether clusters should be shown for the whole network or only for the LCC.</p></li>
<li><p>The <code>print</code> method for <code>summary.microNetProps</code> was completely revised.</p></li>
</ul></div>
<div class="section level4">
<h4 id="plotmicronetprops-1-0-2">plot.microNetProps()<a class="anchor" aria-label="anchor" href="#plotmicronetprops-1-0-2"></a></h4>
<ul><li><p>All normalization methods available for network construction can now be used for scaling node sizes (argument <code>nodeSize</code>).</p></li>
<li><p>New argument <code>normPar</code>: Optional parameters used for normalization.</p></li>
<li><p>Usage of <code>colorVec</code> changed: Node colors can now be set separately in both groups (<code>colorVec</code> can be a single vector or a list with two vectors). Usage depends on <code>nodeColor</code> (see docu of <code>colorVec</code>).</p></li>
<li><p>New argument <code>sameFeatCol</code>: If <code>nodeColor = "feature"</code> and <code>colorVec</code> is not given, <code>sameFeatCol</code> indicates whether same features should have same colors in both groups.</p></li>
<li><p>Argument <code>colorNegAsso</code> has been renamed to <code>negDiffCol</code>. Using the old name leads to a warning.</p></li>
<li><p>New functionality for using the same layout in both groups (if two networks are plotted). In addition to computing the layout for one group and adopting it for the other group, a union of both layout can be computed and used in both groups so that nodes are placed as optimal as possible equally for both networks. This option is applied via <code>sameLayout = TRUE</code> and <code>layoutGroup = "union"</code>. Many thanks to <a href="https://github.com/muellsen?tab=followers" class="external-link">Christian L. Müller</a> and <a href="https://www.iq.harvard.edu/people/alice-sommer" class="external-link">Alice Sommer</a> for providing the idea and R code for this new feature!</p></li>
</ul></div>
<div class="section level4">
<h4 id="netcompare-1-0-2">netCompare()<a class="anchor" aria-label="anchor" href="#netcompare-1-0-2"></a></h4>
<ul><li>
<p>New arguments for storing association and count matrices of the permuted data into an external file:</p>
<ul><li><code>fileLoadAssoPerm</code></li>
<li><code>fileLoadCountsPerm</code></li>
<li><code>storeAssoPerm</code></li>
<li><code>fileStoreAssoPerm</code></li>
<li><code>storeCountsPerm</code></li>
<li><code>fileStoreCountsPerm</code></li>
</ul></li>
<li><p>New argument <code>returnPermProps</code>: If <code>TRUE</code>, global network properties and the respective absolute group differences of the permuted data are returned.</p></li>
<li><p>New argument <code>returnPermCentr</code>: If <code>TRUE</code>, the computed centrality values and the respective absolute group differences of the permuted data are returned as list with a matrix for each centrality measure.</p></li>
<li><p>The arguments <code>assoPerm</code> and <code>dissPerm</code> are still existent for compatibility with <code>NetCoMi 1.0.1</code> but the former elements <code>assoPerm</code> and <code>dissPerm</code> are not returned anymore (matrices are stored in an external file instead).</p></li>
</ul></div>
<div class="section level4">
<h4 id="createassoperm-1-0-2">createAssoPerm()<a class="anchor" aria-label="anchor" href="#createassoperm-1-0-2"></a></h4>
<p>New function for creating association/dissimilarity matrices for permuted count data. The stored count or association/dissimilarity matrices can then be passed to <code><a href="../reference/netCompare.html">netCompare()</a></code> or <code><a href="../reference/diffnet.html">diffnet()</a></code> to decrease runtime. The function also allows to generate a matrix permuted group labels without computing associations. Using this matrix, <code><a href="../reference/createAssoPerm.html">createAssoPerm()</a></code> furthermore allows to estimate the permutation associations/dissimilarities in blocks (by passing only a subset of the permuted group matrix to <code><a href="../reference/createAssoPerm.html">createAssoPerm()</a></code>).</p>
</div>
<div class="section level4">
<h4 id="summarymicronetcomp-1-0-2">summary.microNetComp()<a class="anchor" aria-label="anchor" href="#summarymicronetcomp-1-0-2"></a></h4>
<p>Summary method has been adapted to the new network properties (analogous to the summary of <code>microNetProps</code> objects, which are returned from <code><a href="../reference/netAnalyze.html">netAnalyze()</a></code>)</p>
</div>
<div class="section level4">
<h4 id="diffnet-1-0-2">diffnet()<a class="anchor" aria-label="anchor" href="#diffnet-1-0-2"></a></h4>
<ul><li>
<p>New arguments for storing association and count matrices of the permuted data into an external file:</p>
<ul><li><code>fileLoadAssoPerm</code></li>
<li><code>fileLoadCountsPerm</code></li>
<li><code>storeAssoPerm</code></li>
<li><code>fileStoreAssoPerm</code></li>
<li><code>storeCountsPerm</code></li>
<li><code>fileStoreCountsPerm</code></li>
</ul></li>
<li><p>The argument <code>assoPerm</code> is still existent for compatibility with <code>NetCoMi 1.0.1</code> but the former element <code>assoPerm</code> is not returned anymore (matrices are stored in an external file instead).</p></li>
<li><p>Changed output: For permutation tests and Fisher’s z-test, a vector and matrix with p-values and the corresponding matrix with group differences are returned for both with and without multiple testing adjustment.</p></li>
<li><p>Documentation has been revised.</p></li>
</ul></div>
<div class="section level4">
<h4 id="plotdiffnet-1-0-2">plot.diffnet()<a class="anchor" aria-label="anchor" href="#plotdiffnet-1-0-2"></a></h4>
<ul><li><p>New argument <code>adjusted</code>: Indicates whether the adjacency matrix (matrix with group differences) based on adjusted or unadjusted p-values should be plotted.</p></li>
<li><p>New argument <code>legendPos</code> for positioning the legend.</p></li>
<li><p>New argument <code>legendArgs</code> for specifying further arguments passed to <code>legend</code>.</p></li>
</ul></div>
<div class="section level4">
<h4 id="coltotransp-1-0-2">colToTransp()<a class="anchor" aria-label="anchor" href="#coltotransp-1-0-2"></a></h4>
<ul><li>The function is now exported and its name has changed from <code>col_to_transp()</code> to <code><a href="../reference/colToTransp.html">colToTransp()</a></code>. The function expects a color vector as input and adds transparency to each color.</li>
</ul></div>
</div>
<div class="section level3">
<h3 id="bug-fixes-1-0-2">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-0-2"></a></h3>
<p>The major issues fixed in this release are:</p>
<ul><li><p>The following error is solved: <code>Error in update.list(...): argument "new" is  missing</code>. The error was caused by a conflict between <code>SpiecEasi</code> and <code>metagenomeSeq</code>, in particular by <code>gplot</code> as a dependency of <code>metagenomeSeq</code>. A former version of <code>gplot</code> was dependend on <code>gdata</code>, which caused the conflict. So, please update <code>gplot</code> and remove the package <code>gdata</code> to fix the error.</p></li>
<li><p><code><a href="https://rdrr.io/pkg/SpiecEasi/man/sparcc.html" class="external-link">sparcc()</a></code> from SpiecEasi package is now used for estimating SparCC associations. For some users, NetCoMi’s <code>Rccp</code> implementation of SparCC caused errors when installing NetCoMi. If these are fixed, the Rcpp implementation will be included again, so that users can decide between the two SparCC versions.</p></li>
<li><p>VST transformations are now computed correctly.</p></li>
<li><p>Error when plotting two networks, where one network is empty, has been fixed.</p></li>
</ul></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://stefpeschel.de/" class="external-link">Stefanie Peschel</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

