<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Get started • NetCoMi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get started">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">NetCoMi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/NetCoMi.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/net_comparison.html">Network comparison</a></li>
    <li><a class="dropdown-item" href="../articles/diffnet.html">Differential networks</a></li>
    <li><a class="dropdown-item" href="../articles/dissimilarity_networks.html">Dissimilarity-based Networks</a></li>
    <li><a class="dropdown-item" href="../articles/create_asso_perm.html">Generate permuted association matrices</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>More examples</h6></li>
    <li><a class="dropdown-item" href="../articles/plot_modifications.html">Modifying the network plot</a></li>
    <li><a class="dropdown-item" href="../articles/asso_as_input.html">Association matrix as input</a></li>
    <li><a class="dropdown-item" href="../articles/cross_domain_networks.html">Cross-domain networks</a></li>
    <li><a class="dropdown-item" href="../articles/soil_example.html">Soil microbiome example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/stefpeschel/NetCoMi/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Get started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stefpeschel/NetCoMi/blob/main/vignettes/NetCoMi.Rmd" class="external-link"><code>vignettes/NetCoMi.Rmd</code></a></small>
      <div class="d-none name"><code>NetCoMi.Rmd</code></div>
    </div>

    
    
<p>The American Gut data provided by the <a href="https://github.com/zdk123/SpiecEasi" class="external-link"><code>SpiecEasi</code></a>
package are used for almost all NetCoMi tutorials.</p>
<p>We begin by constructing a single network on genus level, which we
analyze using quantitative and graphical methods. Later, we will compare
the networks of two groups: Individuals with and without seasonal
allergies.</p>
<p>NetCoMi’s main functions are <code><a href="../reference/netConstruct.html">netConstruct()</a></code> for network
construction, <code><a href="../reference/netAnalyze.html">netAnalyze()</a></code> for network analysis, and
<code><a href="../reference/netCompare.html">netCompare()</a></code> for network comparison. As should become clear
from the following examples, these three functions must be executed in
the aforementioned order. A further function is <code><a href="../reference/diffnet.html">diffnet()</a></code>
for constructing a differential association network.
<code><a href="../reference/diffnet.html">diffnet()</a></code> must be applied to the object returned by
<code><a href="../reference/netConstruct.html">netConstruct()</a></code>.</p>
<p>First of all, we load NetCoMi and the data from American Gut Project
(provided by <a href="https://github.com/zdk123/SpiecEasi" class="external-link"><code>SpiecEasi</code></a>,
which is automatically loaded together with NetCoMi).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://netcomi.de">NetCoMi</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load data sets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"amgut1.filt"</span><span class="op">)</span> <span class="co"># ASV count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"amgut2.filt.phy"</span><span class="op">)</span> <span class="co"># phyloseq objext</span></span></code></pre></div>
<div class="section level3">
<h3 id="data-agglomeration">Data agglomeration<a class="anchor" aria-label="anchor" href="#data-agglomeration"></a>
</h3>
<p><em>Skip this step if you wish to construct the network at ASV
level</em>.</p>
<p>We start by agglomerating the phyloseq object to genus level, named
“Rank6” in this data set.</p>
<p>The <code>renameTaxa</code> function is used to rename the taxa
according to a defined pattern and to make the genus names unique.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Agglomerate to genus level</span></span>
<span><span class="va">amgut_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_glom.html" class="external-link">tax_glom</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>, taxrank <span class="op">=</span> <span class="st">"Rank6"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename taxonomic table and make Rank6 (genus) unique</span></span>
<span><span class="va">amgut_genus_renamed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renameTaxa.html">renameTaxa</a></span><span class="op">(</span><span class="va">amgut_genus</span>, </span>
<span>                                  pat <span class="op">=</span> <span class="st">"&lt;name&gt;"</span>, </span>
<span>                                  substPat <span class="op">=</span> <span class="st">"&lt;name&gt;_&lt;subst_name&gt;(&lt;subst_R&gt;)"</span>,</span>
<span>                                  numDupli <span class="op">=</span> <span class="st">"Rank6"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Column 7 contains NAs only and is ignored.</span></span></code></pre></div>
<p>Let’s compare the taxa names before and after renaming them.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;        Rank1         Rank2               Rank3                      </span></span>
<span><span class="co">#&gt; 181016 "k__Bacteria" "p__Firmicutes"     "c__Clostridia"            </span></span>
<span><span class="co">#&gt; 191687 "k__Bacteria" "p__Firmicutes"     "c__Clostridia"            </span></span>
<span><span class="co">#&gt; 305760 "k__Bacteria" "p__Proteobacteria" "c__Gammaproteobacteria"   </span></span>
<span><span class="co">#&gt; 326977 "k__Bacteria" "p__Actinobacteria" "c__Actinobacteria (class)"</span></span>
<span><span class="co">#&gt; 28186  "k__Bacteria" "p__Firmicutes"     "c__Bacilli"               </span></span>
<span><span class="co">#&gt; 541301 "k__Bacteria" "p__Bacteroidetes"  "c__Bacteroidia"           </span></span>
<span><span class="co">#&gt;        Rank4                  Rank5                   Rank6               </span></span>
<span><span class="co">#&gt; 181016 "o__Clostridiales"     "f__Ruminococcaceae"    "g__Eubacterium"    </span></span>
<span><span class="co">#&gt; 191687 "o__Clostridiales"     "f__Lachnospiraceae"    "g__Clostridium"    </span></span>
<span><span class="co">#&gt; 305760 "o__Enterobacteriales" "f__Enterobacteriaceae" "g__Escherichia"    </span></span>
<span><span class="co">#&gt; 326977 "o__Bifidobacteriales" "f__Bifidobacteriaceae" "g__Bifidobacterium"</span></span>
<span><span class="co">#&gt; 28186  "o__Lactobacillales"   "f__Enterococcaceae"    "g__Enterococcus"   </span></span>
<span><span class="co">#&gt; 541301 "o__Bacteroidales"     "f__Porphyromonadaceae" "g__Parabacteroides"</span></span>
<span><span class="co">#&gt;        Rank7</span></span>
<span><span class="co">#&gt; 181016 NA   </span></span>
<span><span class="co">#&gt; 191687 NA   </span></span>
<span><span class="co">#&gt; 305760 NA   </span></span>
<span><span class="co">#&gt; 326977 NA   </span></span>
<span><span class="co">#&gt; 28186  NA   </span></span>
<span><span class="co">#&gt; 541301 NA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus_renamed</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;        Rank1      Rank2            Rank3                    Rank4              </span></span>
<span><span class="co">#&gt; 181016 "Bacteria" "Firmicutes"     "Clostridia"             "Clostridiales"    </span></span>
<span><span class="co">#&gt; 191687 "Bacteria" "Firmicutes"     "Clostridia"             "Clostridiales"    </span></span>
<span><span class="co">#&gt; 305760 "Bacteria" "Proteobacteria" "Gammaproteobacteria"    "Enterobacteriales"</span></span>
<span><span class="co">#&gt; 326977 "Bacteria" "Actinobacteria" "Actinobacteria (class)" "Bifidobacteriales"</span></span>
<span><span class="co">#&gt; 28186  "Bacteria" "Firmicutes"     "Bacilli"                "Lactobacillales"  </span></span>
<span><span class="co">#&gt; 541301 "Bacteria" "Bacteroidetes"  "Bacteroidia"            "Bacteroidales"    </span></span>
<span><span class="co">#&gt;        Rank5                Rank6             Rank7</span></span>
<span><span class="co">#&gt; 181016 "Ruminococcaceae"    "Eubacterium1"    NA   </span></span>
<span><span class="co">#&gt; 191687 "Lachnospiraceae"    "Clostridium1"    NA   </span></span>
<span><span class="co">#&gt; 305760 "Enterobacteriaceae" "Escherichia"     NA   </span></span>
<span><span class="co">#&gt; 326977 "Bifidobacteriaceae" "Bifidobacterium" NA   </span></span>
<span><span class="co">#&gt; 28186  "Enterococcaceae"    "Enterococcus"    NA   </span></span>
<span><span class="co">#&gt; 541301 "Porphyromonadaceae" "Parabacteroides" NA</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxtab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus</span><span class="op">)</span></span>
<span><span class="va">taxtab</span><span class="op">[</span><span class="va">taxtab</span><span class="op">[</span>, <span class="st">"Rank6"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"g__Eubacterium"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [2 taxa by 7 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;        Rank1         Rank2           Rank3           Rank4             </span></span>
<span><span class="co">#&gt; 181016 "k__Bacteria" "p__Firmicutes" "c__Clostridia" "o__Clostridiales"</span></span>
<span><span class="co">#&gt; 189396 "k__Bacteria" "p__Firmicutes" "c__Clostridia" "o__Clostridiales"</span></span>
<span><span class="co">#&gt;        Rank5                Rank6            Rank7</span></span>
<span><span class="co">#&gt; 181016 "f__Ruminococcaceae" "g__Eubacterium" NA   </span></span>
<span><span class="co">#&gt; 189396 "f__Lachnospiraceae" "g__Eubacterium" NA</span></span></code></pre></div>
<p>For example, a “1” has been added to Eubacterium because this genus
exists twice: once in the Ruminococcaceae family and once in the
Lachnospiraceae family. For association estimation, it is important to
distinguish between them, so they are numbered.</p>
</div>
<div class="section level3">
<h3 id="network-construction">Network construction<a class="anchor" aria-label="anchor" href="#network-construction"></a>
</h3>
<p>We use the <a href="https://github.com/GraceYoon/SPRING" class="external-link">SPRING</a>
package for estimating associations (conditional dependence) between
taxa.</p>
<p>The data are filtered within <code><a href="../reference/netConstruct.html">netConstruct()</a></code> as
follows:</p>
<ul>
<li>Only samples with a total number of reads of at least 1000 are
included (argument <code>filtSamp</code>).</li>
<li>Only the 50 taxa with highest frequency are included (argument
<code>filtTax</code>).</li>
</ul>
<p>Note that the taxa filter is set for demonstration purposes only, but
has no effect here because there are only 43 genera in the data set.</p>
<p><code>measure</code> defines the association or dissimilarity
measure, which is <code>"spring"</code> in our case. Additional
arguments are passed to <code><a href="https://rdrr.io/pkg/SPRING/man/SPRING.html" class="external-link">SPRING()</a></code> via
<code>measurePar</code>. <code>nlambda</code> and <code>rep.num</code>
are set to 10 for a decreased execution time, but should be higher for
real data. <code>Rmethod</code> is set to “approx” to estimate the
correlations using a hybrid multi-linear interpolation approach proposed
by <span class="citation">@yoon2020fast</span>. This method considerably
reduces the runtime while controlling the approximation error.</p>
<p>Normalization as well as zero handling is performed internally in
<code><a href="https://rdrr.io/pkg/SPRING/man/SPRING.html" class="external-link">SPRING()</a></code>. Hence, we set <code>normMethod</code> and
<code>zeroMethod</code> to <code>"none"</code>.</p>
<p>We furthermore set <code>sparsMethod</code> to <code>"none"</code>
because <code>SPRING</code> returns a sparse network where no additional
sparsification step is necessary.</p>
<p>We use the “signed” method for transforming associations into
dissimilarities (argument <code>dissFunc</code>). In doing so, strongly
negatively associated taxa have a high dissimilarity and, in turn, a low
similarity, which corresponds to edge weights in the network plot.</p>
<p>The <code>verbose</code> argument is set to 3 so that all messages
generated by <code><a href="../reference/netConstruct.html">netConstruct()</a></code> as well as messages of external
functions are printed.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_spring</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut_genus_renamed</span>,</span>
<span>                           taxRank <span class="op">=</span> <span class="st">"Rank6"</span>,</span>
<span>                           filtTax <span class="op">=</span> <span class="st">"highestFreq"</span>,</span>
<span>                           filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>highestFreq <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>                           filtSamp <span class="op">=</span> <span class="st">"totalReads"</span>,</span>
<span>                           filtSampPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>totalReads <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>,</span>
<span>                           measure <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>                           measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nlambda<span class="op">=</span><span class="fl">10</span>, </span>
<span>                                             rep.num<span class="op">=</span><span class="fl">10</span>,</span>
<span>                                             Rmethod <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span>,</span>
<span>                           normMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           zeroMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>                           sparsMethod <span class="op">=</span> <span class="st">"none"</span>, </span>
<span>                           dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                           seed <span class="op">=</span> <span class="fl">123456</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking input arguments ... Done.</span></span>
<span><span class="co">#&gt; Data filtering ...</span></span>
<span><span class="co">#&gt; 35 samples removed.</span></span>
<span><span class="co">#&gt; 43 taxa and 261 samples remaining.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Calculate 'spring' associations ... Registered S3 method overwritten by 'dendextend':</span></span>
<span><span class="co">#&gt;   method     from </span></span>
<span><span class="co">#&gt;   rev.hclust vegan</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span></span>
<span><span class="co">#&gt;   method         from </span></span>
<span><span class="co">#&gt;   reorder.hclust vegan</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>Let’s take a look at the edge list, which contains for each pair of
nodes the estimated association, the dissimilarity and the adjacency (=
edge weight for weighted networks).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">net_spring</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">)</span></span>
<span><span class="co">#&gt;             v1           v2         asso      diss      adja</span></span>
<span><span class="co">#&gt; 1 Eubacterium1        Dorea  0.003933672 0.7057146 0.2942854</span></span>
<span><span class="co">#&gt; 2 Eubacterium1      Pantoea -0.014829516 0.7123305 0.2876695</span></span>
<span><span class="co">#&gt; 3 Clostridium1 Eubacterium2  0.095333048 0.6725574 0.3274426</span></span>
<span><span class="co">#&gt; 4 Clostridium1   Raoultella -0.012351874 0.7114604 0.2885396</span></span>
<span><span class="co">#&gt; 5 Clostridium1        Dorea  0.015673519 0.7015435 0.2984565</span></span>
<span><span class="co">#&gt; 6 Clostridium1      Cedecea -0.002316558 0.7079253 0.2920747</span></span></code></pre></div>
<p>We can also use NetCoMi’s <code>plotHeat</code> function to plot the
estimated associations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeat.html">plotHeat</a></span><span class="op">(</span><span class="va">net_spring</span><span class="op">$</span><span class="va">assoMat1</span>, textUpp <span class="op">=</span> <span class="st">"none"</span>, textLow <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/unnamed-chunk-6-1.png" width="1152"></p>
</div>
<div class="section level3">
<h3 id="network-analysis">Network analysis<a class="anchor" aria-label="anchor" href="#network-analysis"></a>
</h3>
<p>NetCoMi’s <code><a href="../reference/netAnalyze.html">netAnalyze()</a></code> function is used for analyzing
the constructed network.</p>
<p>Here, <code>centrLCC</code> is set to <code>TRUE</code> meaning that
centralities are calculated only for nodes in the largest connected
component (LCC).</p>
<p>Clusters are identified using greedy modularity optimization (by
<code>cluster_fast_greedy()</code> from <a href="https://igraph.org/r/" class="external-link"><code>igraph</code></a> package).</p>
<p>Hubs are nodes with an eigenvector centrality value above the
empirical 95% quantile of all eigenvector centralities in the network
(argument <code>hubPar</code>).</p>
<p><code>weightDeg</code> and <code>normDeg</code> are set to
<code>FALSE</code> so that the degree of a node is simply defined as
number of nodes that are adjacent to the node.</p>
<p>By default, a heatmap of the Graphlet Correlation Matrix (GCM) is
returned (with graphlet correlations in the upper triangle and
significance codes resulting from Student’s t-test in the lower
triangle). See <code><a href="../reference/calcGCM.html">?calcGCM</a></code> and <code><a href="../reference/testGCM.html">?testGCM</a></code> for
details.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_spring</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_spring</span>, </span>
<span>                           centrLCC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                           hubPar <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>                           weightDeg <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           normDeg <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_spring_2-1.png" width="576"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#?summary.microNetProps</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">props_spring</span>, numbNodes <span class="op">=</span> <span class="fl">5L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Component sizes</span></span>
<span><span class="co">#&gt; ```````````````            </span></span>
<span><span class="co">#&gt; size: 40 2 1</span></span>
<span><span class="co">#&gt;    #:  1 1 1</span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Global network properties</span></span>
<span><span class="co">#&gt; `````````````````````````</span></span>
<span><span class="co">#&gt; Largest connected component (LCC):</span></span>
<span><span class="co">#&gt;                                  </span></span>
<span><span class="co">#&gt; Relative LCC size         0.93023</span></span>
<span><span class="co">#&gt; Clustering coefficient    0.33039</span></span>
<span><span class="co">#&gt; Modularity                0.41972</span></span>
<span><span class="co">#&gt; Positive edge percentage 83.01887</span></span>
<span><span class="co">#&gt; Edge density              0.13590</span></span>
<span><span class="co">#&gt; Natural connectivity      0.03685</span></span>
<span><span class="co">#&gt; Vertex connectivity       1.00000</span></span>
<span><span class="co">#&gt; Edge connectivity         1.00000</span></span>
<span><span class="co">#&gt; Average dissimilarity*    0.95654</span></span>
<span><span class="co">#&gt; Average path length**     1.84125</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Whole network:</span></span>
<span><span class="co">#&gt;                                  </span></span>
<span><span class="co">#&gt; Number of components      3.00000</span></span>
<span><span class="co">#&gt; Clustering coefficient    0.33039</span></span>
<span><span class="co">#&gt; Modularity                0.42746</span></span>
<span><span class="co">#&gt; Positive edge percentage 82.24299</span></span>
<span><span class="co">#&gt; Edge density              0.11849</span></span>
<span><span class="co">#&gt; Natural connectivity      0.03362</span></span>
<span><span class="co">#&gt; -----</span></span>
<span><span class="co">#&gt; *: Dissimilarity = 1 - edge weight</span></span>
<span><span class="co">#&gt; **: Path length = Units with average dissimilarity</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Clusters</span></span>
<span><span class="co">#&gt; - In the whole network</span></span>
<span><span class="co">#&gt; - Algorithm: cluster_fast_greedy</span></span>
<span><span class="co">#&gt; ```````````````````````````````` </span></span>
<span><span class="co">#&gt;                    </span></span>
<span><span class="co">#&gt; name: 0  1 2  3 4 5</span></span>
<span><span class="co">#&gt;    #: 1 10 9 12 2 9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Hubs</span></span>
<span><span class="co">#&gt; - In alphabetical/numerical order</span></span>
<span><span class="co">#&gt; - Based on empirical quantiles of centralities</span></span>
<span><span class="co">#&gt; ```````````````````````````````````````````````                        </span></span>
<span><span class="co">#&gt;  5_Clostridiales(O)     </span></span>
<span><span class="co">#&gt;  6_Enterobacteriaceae(F)</span></span>
<span><span class="co">#&gt;  Clostridium1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Centrality measures</span></span>
<span><span class="co">#&gt; - In decreasing order</span></span>
<span><span class="co">#&gt; - Centrality of disconnected components is zero</span></span>
<span><span class="co">#&gt; ````````````````````````````````````````````````</span></span>
<span><span class="co">#&gt; Degree (unnormalized):</span></span>
<span><span class="co">#&gt;                           </span></span>
<span><span class="co">#&gt; Dorea                   11</span></span>
<span><span class="co">#&gt; 5_Clostridiales(O)      11</span></span>
<span><span class="co">#&gt; 6_Enterobacteriaceae(F) 11</span></span>
<span><span class="co">#&gt; Clostridium1            10</span></span>
<span><span class="co">#&gt; Raoultella              10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Betweenness centrality (normalized):</span></span>
<span><span class="co">#&gt;                             </span></span>
<span><span class="co">#&gt; 2_Ruminococcaceae(F) 0.17274</span></span>
<span><span class="co">#&gt; Dorea                0.16599</span></span>
<span><span class="co">#&gt; Odoribacter          0.13495</span></span>
<span><span class="co">#&gt; 5_Clostridiales(O)   0.12686</span></span>
<span><span class="co">#&gt; Faecalibacterium     0.12551</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Closeness centrality (normalized):</span></span>
<span><span class="co">#&gt;                                </span></span>
<span><span class="co">#&gt; 5_Clostridiales(O)      0.83523</span></span>
<span><span class="co">#&gt; Clostridium1            0.82623</span></span>
<span><span class="co">#&gt; 2_Ruminococcaceae(F)    0.81965</span></span>
<span><span class="co">#&gt; Faecalibacterium        0.81242</span></span>
<span><span class="co">#&gt; 6_Enterobacteriaceae(F) 0.81208</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvector centrality (normalized):</span></span>
<span><span class="co">#&gt;                                </span></span>
<span><span class="co">#&gt; 5_Clostridiales(O)      1.00000</span></span>
<span><span class="co">#&gt; Clostridium1            0.99952</span></span>
<span><span class="co">#&gt; 6_Enterobacteriaceae(F) 0.93828</span></span>
<span><span class="co">#&gt; Dorea                   0.88613</span></span>
<span><span class="co">#&gt; 3_Lachnospiraceae(F)    0.81654</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualizing-the-network">Visualizing the network<a class="anchor" aria-label="anchor" href="#visualizing-the-network"></a>
</h3>
<p>We use the determined clusters as node colors and scale the node
sizes according to the node’s eigenvector centrality.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># help page</span></span>
<span><span class="op">?</span><span class="va">plot.microNetProps</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_spring</span>, </span>
<span>          labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>          nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>          title1 <span class="op">=</span> <span class="st">"Network on genus level with SPRING associations"</span>, </span>
<span>          showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          cexTitle <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>          cexLabels <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated association:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_spring_3-1.png" width="1920"></p>
<p>Note that edge weights are (non-negative) similarities, however, the
edges belonging to negative estimated associations are colored in red by
default (<code>negDiffCol = TRUE</code>).</p>
<p>By default, a different transparency value is added to edges with an
absolute weight below and above the <code>cut</code> value (arguments
<code>edgeTranspLow</code> and <code>edgeTranspHigh</code>). The
determined <code>cut</code> value can be read out as follows:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">q1</span><span class="op">$</span><span class="va">Arguments</span><span class="op">$</span><span class="va">cut</span></span>
<span><span class="co">#&gt;       75% </span></span>
<span><span class="co">#&gt; 0.3271051</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="export-to-gephi">Export to Gephi<a class="anchor" aria-label="anchor" href="#export-to-gephi"></a>
</h3>
<p>Some users may be interested in how to export the network to Gephi.
Here’s an example:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For Gephi, we have to generate an edge list with IDs.</span></span>
<span><span class="co"># The corresponding labels (and also further node features) are stored as node list.</span></span>
<span></span>
<span><span class="co"># Create edge object from the edge list exported by netConstruct()</span></span>
<span><span class="va">edges</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">net_spring</span><span class="op">$</span><span class="va">edgelist1</span>, <span class="va">v1</span>, <span class="va">v2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add Source and Target variables (as IDs)</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Source</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">edges</span><span class="op">$</span><span class="va">v1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Target</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">edges</span><span class="op">$</span><span class="va">v2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Type</span> <span class="op">&lt;-</span> <span class="st">"Undirected"</span></span>
<span><span class="va">edges</span><span class="op">$</span><span class="va">Weight</span> <span class="op">&lt;-</span> <span class="va">net_spring</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">$</span><span class="va">adja</span></span>
<span></span>
<span><span class="va">nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">edges</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'v1'</span>,<span class="st">'Source'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Label"</span>, <span class="st">"Id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add category with clusters (can be used as node colors in Gephi)</span></span>
<span><span class="va">nodes</span><span class="op">$</span><span class="va">Category</span> <span class="op">&lt;-</span> <span class="va">props_spring</span><span class="op">$</span><span class="va">clustering</span><span class="op">$</span><span class="va">clust1</span><span class="op">[</span><span class="va">nodes</span><span class="op">$</span><span class="va">Label</span><span class="op">]</span></span>
<span></span>
<span><span class="va">edges</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">edges</span>, <span class="va">Source</span>, <span class="va">Target</span>, <span class="va">Type</span>, <span class="va">Weight</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">nodes</span>, file <span class="op">=</span> <span class="st">"nodes.csv"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">edges</span>, file <span class="op">=</span> <span class="st">"edges.csv"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The exported .csv files can then be imported into Gephi.</p>
<hr>
</div>
<div class="section level3">
<h3 id="network-with-pearson-correlation-as-association-measure">Network with Pearson correlation as association measure<a class="anchor" aria-label="anchor" href="#network-with-pearson-correlation-as-association-measure"></a>
</h3>
<p>Let’s construct another network using Pearson’s correlation
coefficient as association measure. The input is now a
<code>phyloseq</code> object.</p>
<p>Since Pearson correlations may lead to compositional effects when
applied to sequencing data, we use the clr transformation as
normalization method. Zero treatment is necessary in this case.</p>
<p>A threshold of 0.3 is used as sparsification method, so that only
OTUs with an absolute correlation greater than or equal to 0.3 are
connected.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_pears</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>,  </span>
<span>                          measure <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>                          normMethod <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                          zeroMethod <span class="op">=</span> <span class="st">"multRepl"</span>,</span>
<span>                          sparsMethod <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>                          thresh <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking input arguments ... Done.</span></span>
<span><span class="co">#&gt; 2 rows with zero sum removed.</span></span>
<span><span class="co">#&gt; 138 taxa and 294 samples remaining.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Zero treatment:</span></span>
<span><span class="co">#&gt; Execute multRepl() ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Normalization:</span></span>
<span><span class="co">#&gt; Execute clr(){SpiecEasi} ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Calculate 'pearson' associations ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Sparsify associations via 'threshold' ... Done.</span></span></code></pre></div>
<p>Network analysis and plotting:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_pears</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_pears</span>, </span>
<span>                          clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_2-1.png" width="576"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on OTU level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>, </span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_3-1.png" width="1920"></p>
<p>Let’s improve the visualization by changing the following
arguments:</p>
<ul>
<li>
<code>repulsion = 0.8</code>: Place the nodes further apart.</li>
<li>
<code>rmSingles = TRUE</code>: Single nodes are removed.</li>
<li>
<code>labelScale = FALSE</code> and <code>cexLabels = 1.6</code>:
All labels have equal size and are enlarged to improve readability of
small node’s labels.</li>
<li>
<code>nodeSizeSpread = 3</code> (default is 4): Node sizes are more
similar if the value is decreased. This argument (in combination with
<code>cexNodes</code>) is useful to enlarge small nodes while keeping
the size of big nodes.</li>
<li>
<code>hubBorderCol = "darkgray"</code>: Change border color for a
better readability of the node labels.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.6</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on OTU level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_4-1.png" width="1920"></p>
<div class="section level4">
<h4 id="edge-filtering">Edge filtering<a class="anchor" aria-label="anchor" href="#edge-filtering"></a>
</h4>
<p>The network can be sparsified further using the arguments
<code>edgeFilter</code> (edges are filtered before the layout is
computed) and <code>edgeInvisFilter</code> (edges are removed after the
layout is computed and thus just made “invisible”).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears</span>,</span>
<span>     edgeInvisFilter <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>     edgeInvisPar <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.6</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Network on OTU level with Pearson correlations"</span>,</span>
<span>                     <span class="st">"\n(edge filter: threshold = 0.4)"</span><span class="op">)</span>,</span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_5-1.png" width="1920"></p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="using-the-unsigned-transformation">Using the “unsigned” transformation<a class="anchor" aria-label="anchor" href="#using-the-unsigned-transformation"></a>
</h3>
<p>In the above network, the “signed” transformation was used to
transform the estimated associations into dissimilarities. This leads to
a network where strongly positive correlated taxa have a high edge
weight (1 if the correlation equals 1) and strongly negative correlated
taxa have a low edge weight (0 if the correlation equals -1).</p>
<p>We now use the “unsigned” transformation so that the edge weight
between strongly correlated taxa is high, no matter of the sign. Hence,
a correlation of -1 and 1 would lead to an edge weight of 1.</p>
<div class="section level4">
<h4 id="network-construction-1">Network construction<a class="anchor" aria-label="anchor" href="#network-construction-1"></a>
</h4>
<p>We can pass the network object from before to
<code><a href="../reference/netConstruct.html">netConstruct()</a></code> to save runtime.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net_pears_unsigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netConstruct.html">netConstruct</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">net_pears</span><span class="op">$</span><span class="va">assoEst1</span>,</span>
<span>                                   dataType <span class="op">=</span> <span class="st">"correlation"</span>, </span>
<span>                                   sparsMethod <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>                                   thresh <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                                   dissFunc <span class="op">=</span> <span class="st">"unsigned"</span>,</span>
<span>                                   verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking input arguments ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Sparsify associations via 'threshold' ... Done.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="estimated-correlations-and-adjacency-values">Estimated correlations and adjacency values<a class="anchor" aria-label="anchor" href="#estimated-correlations-and-adjacency-values"></a>
</h4>
<p>The following histograms demonstrate how the estimated correlations
are transformed into adjacencies (= sparsified similarities for weighted
networks).</p>
<p>Sparsified estimated correlations:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">net_pears</span><span class="op">$</span><span class="va">assoMat1</span>, <span class="fl">100</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Estimated correlation"</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Estimated correlations after sparsification"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_hist_1-1.png" width="768"></p>
<p>Adjacency values computed using the “signed” transformation (values
different from 0 and 1 will be edges in the network):</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">net_pears</span><span class="op">$</span><span class="va">adjaMat1</span>, <span class="fl">100</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Adjacency values"</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Adjacencies (with \"signed\" transformation)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_hist_2-1.png" width="768"></p>
<p>Adjacency values computed using the “unsigned” transformation:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">net_pears_unsigned</span><span class="op">$</span><span class="va">adjaMat1</span>, <span class="fl">100</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">400</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Adjacency values"</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Adjacencies (with \"unsigned\" transformation)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_hist_3-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="network-analysis-and-plotting">Network analysis and plotting<a class="anchor" aria-label="anchor" href="#network-analysis-and-plotting"></a>
</h4>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">props_pears_unsigned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_pears_unsigned</span>, </span>
<span>                                   clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                                   gcmHeat <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_pears_unsigned</span>, </span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.6</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network with Pearson correlations and \"unsigned\" transformation"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_pears_unsigned_4-1.png" width="1920"></p>
<p>While with the “signed” transformation, positive correlated taxa are
likely to belong to the same cluster, with the “unsigned” transformation
clusters contain strongly positive and negative correlated taxa.</p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="network-on-genus-level">Network on genus level<a class="anchor" aria-label="anchor" href="#network-on-genus-level"></a>
</h3>
<p>We now construct a further network, where OTUs are agglomerated to
genera.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"amgut2.filt.phy"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Agglomerate to genus level</span></span>
<span><span class="va">amgut_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_glom.html" class="external-link">tax_glom</a></span><span class="op">(</span><span class="va">amgut2.filt.phy</span>, taxrank <span class="op">=</span> <span class="st">"Rank6"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Taxonomic table</span></span>
<span><span class="va">taxtab</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename taxonomic table and make Rank6 (genus) unique</span></span>
<span><span class="va">amgut_genus_renamed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renameTaxa.html">renameTaxa</a></span><span class="op">(</span><span class="va">amgut_genus</span>, </span>
<span>                                  pat <span class="op">=</span> <span class="st">"&lt;name&gt;"</span>, </span>
<span>                                  substPat <span class="op">=</span> <span class="st">"&lt;name&gt;_&lt;subst_name&gt;(&lt;subst_R&gt;)"</span>,</span>
<span>                                  numDupli <span class="op">=</span> <span class="st">"Rank6"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Column 7 contains NAs only and is ignored.</span></span>
<span></span>
<span><span class="co"># Network construction and analysis</span></span>
<span><span class="va">net_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netConstruct.html">netConstruct</a></span><span class="op">(</span><span class="va">amgut_genus_renamed</span>,</span>
<span>                          taxRank <span class="op">=</span> <span class="st">"Rank6"</span>,</span>
<span>                          measure <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>                          zeroMethod <span class="op">=</span> <span class="st">"multRepl"</span>,</span>
<span>                          normMethod <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>                          sparsMethod <span class="op">=</span> <span class="st">"threshold"</span>,</span>
<span>                          thresh <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking input arguments ...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; 2 rows with zero sum removed.</span></span>
<span><span class="co">#&gt; 43 taxa and 294 samples remaining.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Zero treatment:</span></span>
<span><span class="co">#&gt; Execute multRepl() ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Normalization:</span></span>
<span><span class="co">#&gt; Execute clr(){SpiecEasi} ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Calculate 'pearson' associations ... Done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Sparsify associations via 'threshold' ... Done.</span></span>
<span></span>
<span><span class="va">props_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/netAnalyze.html">netAnalyze</a></span><span class="op">(</span><span class="va">net_genus</span>, clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_genus_1-1.png" width="576"></p>
<div class="section level4">
<h4 id="network-plots">Network plots<a class="anchor" aria-label="anchor" href="#network-plots"></a>
</h4>
<p>Modifications:</p>
<ul>
<li>Fruchterman-Reingold layout algorithm from <code>igraph</code>
package used (passed to <code>plot</code> as matrix)</li>
<li>Shortened labels (using the “intelligent” method, which avoids
duplicates)</li>
<li>Fixed node sizes, where hubs are enlarged</li>
<li>Node color is gray for all nodes (transparancy is lower for hub
nodes by default)</li>
</ul>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute layout</span></span>
<span><span class="va">graph3</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_adjacency_matrix.html" class="external-link">graph_from_adjacency_matrix</a></span><span class="op">(</span><span class="va">net_genus</span><span class="op">$</span><span class="va">adjaMat1</span>, </span>
<span>                                              weighted <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span><span class="va">lay_fr</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/layout_with_fr.html" class="external-link">layout_with_fr</a></span><span class="op">(</span><span class="va">graph3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Row names of the layout matrix must match the node names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lay_fr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">net_genus</span><span class="op">$</span><span class="va">adjaMat1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_genus</span>,</span>
<span>     layout <span class="op">=</span> <span class="va">lay_fr</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"intelligent"</span>,</span>
<span>     labelLength <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     labelPattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"'"</span>, <span class="fl">3</span>, <span class="st">"'"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"fix"</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"gray"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     cexHubs <span class="op">=</span> <span class="fl">1.1</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on genus level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_genus_2-1.png" width="1920"></p>
<p>Since the above visualization is obviously not optimal, we make
further adjustments:</p>
<ul>
<li>This time, the Fruchterman-Reingold layout algorithm is computed
within the plot function and thus applied to the “reduced” network
without singletons</li>
<li>Labels are not scaled to node sizes</li>
<li>Single nodes are removed</li>
<li>Node sizes are scaled to the column sums of clr-transformed
data</li>
<li>Node colors represent the determined clusters</li>
<li>Border color of hub nodes is changed from black to darkgray</li>
<li>Label size of hubs is enlarged</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_genus</span>,</span>
<span>     layout <span class="op">=</span> <span class="st">"layout_with_fr"</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"intelligent"</span>,</span>
<span>     labelLength <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     labelPattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"'"</span>, <span class="fl">3</span>, <span class="st">"'"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"darkgray"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on genus level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009900"</span>,<span class="st">"red"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_genus_3-1.png" width="1920"></p>
<p>Let’s check whether the largest nodes are actually those with highest
column sums in the matrix with normalized counts returned by
<code><a href="../reference/netConstruct.html">netConstruct()</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">net_genus</span><span class="op">$</span><span class="va">normCounts1</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;             Bacteroides              Klebsiella        Faecalibacterium </span></span>
<span><span class="co">#&gt;               1200.7971               1137.4928                708.0877 </span></span>
<span><span class="co">#&gt;      5_Clostridiales(O)    2_Ruminococcaceae(F)    3_Lachnospiraceae(F) </span></span>
<span><span class="co">#&gt;                549.2647                502.1889                493.7558 </span></span>
<span><span class="co">#&gt; 6_Enterobacteriaceae(F)               Roseburia         Parabacteroides </span></span>
<span><span class="co">#&gt;                363.3841                333.8737                328.0495 </span></span>
<span><span class="co">#&gt;             Coprococcus </span></span>
<span><span class="co">#&gt;                274.4082</span></span></code></pre></div>
<p>In order to further improve our plot, we use the following
modifications:</p>
<ul>
<li>This time, we choose the “spring” layout as part of
<code>qgraph()</code> (the function is generally used for network
plotting in NetCoMi)</li>
<li>A repulsion value below 1 places the nodes further apart</li>
<li>Labels are not shortened anymore</li>
<li>Nodes (bacteria on genus level) are colored according to the
respective phylum</li>
<li>Edges representing positive associations are colored in blue,
negative ones in orange (just to give an example for alternative edge
coloring)</li>
<li>Transparency is increased for edges with high weight to improve the
readability of node labels</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get phyla names</span></span>
<span><span class="va">taxtab</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">amgut_genus_renamed</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"p__"</span>, <span class="st">""</span>, <span class="va">taxtab</span><span class="op">[</span>, <span class="st">"Rank2"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">taxtab</span><span class="op">[</span>, <span class="st">"Rank6"</span><span class="op">]</span></span>
<span><span class="co">#table(phyla)</span></span>
<span></span>
<span><span class="co"># Define phylum colors</span></span>
<span><span class="va">phylcol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cyan"</span>, <span class="st">"blue3"</span>, <span class="st">"red"</span>, <span class="st">"lawngreen"</span>, <span class="st">"yellow"</span>, <span class="st">"deeppink"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot.microNetProps.html">plot</a></span><span class="op">(</span><span class="va">props_genus</span>,</span>
<span>     layout <span class="op">=</span> <span class="st">"spring"</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">0.84</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     charToRm <span class="op">=</span> <span class="st">"g__"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">4</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"feature"</span>, </span>
<span>     featVecCol <span class="op">=</span> <span class="va">phyla</span>, </span>
<span>     colorVec <span class="op">=</span>  <span class="va">phylcol</span>,</span>
<span>     posCol <span class="op">=</span> <span class="st">"darkturquoise"</span>, </span>
<span>     negCol <span class="op">=</span> <span class="st">"orange"</span>,</span>
<span>     edgeTranspLow <span class="op">=</span> <span class="fl">0</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">40</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexLabels <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     cexHubLabels <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network on genus level with Pearson correlations"</span>, </span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2.3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span><span class="va">phylcol_transp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colToTransp.html">colToTransp</a></span><span class="op">(</span><span class="va">phylcol</span>, <span class="fl">60</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.2</span>, <span class="fl">1.2</span>, cex <span class="op">=</span> <span class="fl">2</span>, pt.cex <span class="op">=</span> <span class="fl">2.5</span>, title <span class="op">=</span> <span class="st">"Phylum:"</span>, </span>
<span>       legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">phylcol_transp</span>, bty <span class="op">=</span> <span class="st">"n"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">2.2</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkturquoise"</span>,<span class="st">"orange"</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="NetCoMi_files/figure-html/single_genus_5-1.png" width="1920"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://stefpeschel.de/" class="external-link">Stefanie Peschel</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
